- **Calling variants in diploid systems**
  - Pro:
    - cartoons are nice (*2019-03-04 00:00:00*)
  - Con:
    - I'm still in the middle of it.  But when doing snpEFF there is no refernece genome available.  This tutorial does not say to download one.  Other variant analyses say to download hg19.  this is not available in the shared data folders.  When using an hg19 fasta from another source freeBayes fails because it isn't indexed.  Not sure how to index it.  SnpEFF calls for a database.  I downloaded one and am trying it now.  But it would be good to include this step in the tutorial (*2018-11-04 00:00:00*)
    - There is a problem in SnpEff step. There is no option for "Locally installed reference genome" when clicking Genome resource causing the next step cannot be proceed (*2019-04-15 00:00:00*)

- **Calling variants in non-diploid systems**
  - Pro:
    - I like the idea (*2018-11-04 00:00:00*)
  - Con:
    - After running the FreeBayes I ended up with 2 variants instead of 30.  And so got stuck there.  Also it's not clear why the filtering and subsequent steps would be done on the markduplicates output, rather than the left align output.  Why do the leftalign if we don't use that output again?  I did both and got the same results. Overall I am finding the galaxy experience to be rather frustrating.  I'm trying to decide whether to incorporate these tutorials in my class, but often I can't find the necessary input data, get unexplained errors, or get output different from the tutorial (*2018-11-04 00:00:00*)

- **Exome sequencing data analysis**
  - Pro:
    - I like the idea of it (*2018-11-04 00:00:00*)
  - Con:
    - It seems impossible to complete.  The training data is not available in the shared data link.  If it is, it's too hard to find.  A more general point would be to explain how the tutorial could be used on other data sets.  or ref genomes.  For this I tried using an hg19 fasta file from another data set, but got an error about indexing.  So I guess, explaining how the training data sets are pre-formatted would help when trying to use sets from other sources (*2018-11-04 00:00:00*)
    - unfortunately in compare with others tutorial, this is very unclear and o'm not satisfied please improve this very important tutorial with clear and accurate dive. (*2018-12-26 00:00:00*)

- **Exome sequencing data analysis for diagnosing a genetic disease**
  - Pro:
    - looking at genes (*2019-07-26 00:00:00*)
    - Even simple steps were explained. (*2019-07-26 00:00:00*)
    - How the tuturial is complete (*2019-07-26 00:00:00*)
  - Con:
    - Remember to list the softwares that is in galaxy because Gemini ins't there (*2019-07-26 00:00:00*)

- **Identification of somatic and germline variants from tumor and normal sample pairs**


- **Mapping and molecular identification of phenotype-causing mutations**
  - Pro:
    - all (*2019-04-09 00:00:00*)
  - Con:
    - more screenshots (*2019-04-09 00:00:00*)

- **Microbial Variant Calling**
  - Pro:
    - This was very straight forward and ran well (once it was clear that it only runs on the eu site).  listing the file types was useful (*2018-11-04 00:00:00*)
    - the tutorial is quiet clear, plus the images and screenshots are really helpful (*2019-09-11 00:00:00*)
    - Hands on work (*2019-09-11 00:00:00*)
  - Con:
    - maybe link to descriptions of the various file type generated/used (*2018-11-04 00:00:00*)
    - The step "Select all outputs" in "Hands-on: Run Snippy" is not obvious what to do exactly.  (*2019-07-24 00:00:00*)
    - Good experience using Galaxy (*2019-09-11 00:00:00*)

