{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Purpose:** Extract the details about the current resources in the Galaxy Training Material and some statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'2019-11-06 10:19:43.065429'"
      ]
     },
     "execution_count": 217,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import base64\n",
    "import copy\n",
    "import datetime\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import re\n",
    "import yaml\n",
    "\n",
    "from github import Github\n",
    "from pprint import pprint\n",
    "\n",
    "\n",
    "str(datetime.datetime.now())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Usage:\n",
    "1. Generate a Personal access tokens on GitHub (in Setting)\n",
    "2. Add it to the `../config.yaml` after `github:`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"../config.yaml\", \"r\") as stream:\n",
    "    config = yaml.load(stream, Loader=yaml.FullLoader)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# connect to GitHub\n",
    "g = Github(config[\"github\"])\n",
    "# retrieve the hub repository\n",
    "repo = g.get_user(\"galaxyproject\").get_repo(\"training-material\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dates and corresponding commit\n",
    "gitcommit_dates = {\n",
    "    \"22.06.17\": \"44164c9b\",\n",
    "    \"22.06.18\": \"24ea24ef\",\n",
    "    \"22.06.19\": \"4ae23d58\",\n",
    "    'today': \"master\"}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Evolution of topics, tutorials and technical support"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [],
   "source": [
    "def has_zenodo(file_content):\n",
    "    '''\n",
    "    Check if Zenodo link in tutorial\n",
    "\n",
    "    :param file_content: the encoded content of the file obtained with GitHub API\n",
    "\n",
    "    :return: boolean\n",
    "    '''\n",
    "    decoded_file_content = base64.b64decode(file_content)\n",
    "    return 'zenodo_link' in str(decoded_file_content)\n",
    "\n",
    "\n",
    "def is_enable(file_content):\n",
    "    '''\n",
    "    Check if tutorial is enabled or not link in tutorial\n",
    "\n",
    "    :param file_content: the encoded content of the file obtained with GitHub API\n",
    "\n",
    "    :return: boolean\n",
    "    '''\n",
    "    decoded_file_content = base64.b64decode(file_content)\n",
    "    return not 'enable: ' in str(decoded_file_content)\n",
    "\n",
    "\n",
    "def get_topic_title(topic_n, ref):\n",
    "    '''\n",
    "    :param topic_n: name of the topic\n",
    "    :param ref: name of the commit/branch/tag\n",
    "    \n",
    "    :return: title of the topic in metadata file\n",
    "    '''\n",
    "    metadata_f = repo.get_contents(\"topics/%s/metadata.yaml\" % topic_n, ref)\n",
    "    metadata_f_content = yaml.load(base64.b64decode(metadata_f.content), Loader=yaml.FullLoader)\n",
    "    return metadata_f_content['title']\n",
    "    \n",
    "\n",
    "def extract_training_content(ref):\n",
    "    '''\n",
    "    Extract the details about the current resources in the Galaxy Training Material,\n",
    "    and aggregate information per topics\n",
    "\n",
    "    :param ref: name of the commit/branch/tag\n",
    "\n",
    "    :return: dictionary with information about the tutorials (tutorial, slides, workflow, \n",
    "    data library) in each topic on ref and aggregated statistics per topics\n",
    "    '''\n",
    "    trainings = {}\n",
    "    topics_dir = \"topics\"\n",
    "    infos = {'slides': False, 'hands_on': False, 'workflows': False, 'data_library': False, 'zenodo': False}\n",
    "    stat_c = {'introduction': 0, 'tutorials': 0, 'slides': 0, 'hands_on': 0, 'workflows': 0, 'zenodo': 0, 'data_library': 0}\n",
    "    stats = pd.DataFrame(columns=list(stat_c.keys()), index=[])\n",
    "    for topic in repo.get_dir_contents(\"topics\", ref):\n",
    "        # get topic details\n",
    "        topic_n = topic.name\n",
    "        topic_t = get_topic_title(topic_n, gitcommit_dates[d])\n",
    "        # prepare aggregation for topic\n",
    "        trainings.setdefault(topic_n, {})\n",
    "        topic_stat = copy.copy(stat_c)\n",
    "        # check for introduction\n",
    "        for topic_c in repo.get_dir_contents(\"%s/\" % topic.path, ref):\n",
    "            if topic_c.name == 'slides':\n",
    "                topic_stat['introduction'] += len(\n",
    "                    repo.get_dir_contents(\"%s/slides/\" % topic.path, ref))\n",
    "        # parse the tutorials\n",
    "        for tuto in repo.get_dir_contents(\"%s/tutorials/\" % topic.path, ref):\n",
    "            tuto_n = tuto.name\n",
    "            enabled = True\n",
    "            trainings[topic_n].setdefault(tuto_n, copy.copy(infos))\n",
    "            # parse the files in the tuto folder\n",
    "            for tuto_f in repo.get_dir_contents(tuto.path, ref):\n",
    "                if tuto_f.name == 'slides.html':\n",
    "                    trainings[topic_n][tuto_n]['slides'] = True\n",
    "                    enabled = is_enable(tuto_f.content)\n",
    "                elif tuto_f.name == 'tutorial.md':\n",
    "                    trainings[topic_n][tuto_n]['hands_on'] = True\n",
    "                    trainings[topic_n][tuto_n]['zenodo'] = has_zenodo(tuto_f.content)\n",
    "                    enabled = is_enable(tuto_f.content)\n",
    "                elif tuto_f.name == 'workflows':\n",
    "                    trainings[topic_n][tuto_n]['workflows'] = True\n",
    "                elif tuto_f.name == 'data-library.yaml':\n",
    "                    trainings[topic_n][tuto_n]['data_library'] = True\n",
    "            # aggregate stats if enable tutorial\n",
    "            if enabled:\n",
    "                topic_stat['tutorials'] += 1\n",
    "                for c in topic_stat:\n",
    "                    if c == 'tutorials' or c == 'introduction':\n",
    "                        continue\n",
    "                    if trainings[topic_n][tuto_n][c]:\n",
    "                        topic_stat[c] += 1\n",
    "        # add stats in table\n",
    "        stats.loc[topic_t] = topic_stat\n",
    "    # rename table columns\n",
    "    stats = stats.rename(columns = {\n",
    "        'introduction': 'Introduction slide decks',\n",
    "        'tutorials': 'Tutorials',\n",
    "        'slides': 'Slide decks',\n",
    "        'hands_on': 'Hands-on tutorials',\n",
    "        'workflows': 'Workflows',\n",
    "        'zenodo': 'Data on Zenodo',\n",
    "        'data_library': 'Data on data library'})         \n",
    "            \n",
    "    return trainings, stats    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# extract training content\n",
    "training_contents = {}\n",
    "stats_per_date = {}\n",
    "global_stats = pd.DataFrame(columns=[], index=[])\n",
    "for d, git_com in gitcommit_dates.items():\n",
    "    training_contents[d], stats_per_date[d] = extract_training_content(git_com)\n",
    "    aggr_stats = (stats_per_date[d]\n",
    "                  .sum()\n",
    "                  .add(pd.Series([len(training_contents[d])], index=['Topics']), fill_value=0)\n",
    "                  .to_frame(name=d))\n",
    "    global_stats = pd.concat([global_stats, aggr_stats], axis=1)\n",
    "# format table\n",
    "global_stats = (global_stats\n",
    "    .reindex(['Topics', 'Introduction slide decks', 'Tutorials', 'Slide decks', 'Hands-on tutorials', 'Workflows', 'Data on Zenodo', 'Data on data library'])\n",
    "    .astype(int))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Topics, number of tutorials and technical support per dates (22.06.17, 22.06.18, 22.06.19, today)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Introduction slides</th>\n",
       "      <th>Tutorials</th>\n",
       "      <th>Slide decks</th>\n",
       "      <th>Hands-on tutorials</th>\n",
       "      <th>Workflows</th>\n",
       "      <th>Data on Zenodo</th>\n",
       "      <th>Data on data library</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Galaxy Server administration</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Assembly</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ChIP-Seq data analysis</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Development in Galaxy</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Epigenetics</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Galaxy Introduction</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Metagenomics</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Proteomics</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Analysis of sequences</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Train the trainers</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Transcriptomics</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Tutorials using public Galaxy at usegalaxy.org</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Variant Analysis</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               Introduction slides Tutorials  \\\n",
       "Galaxy Server administration                                     1         4   \n",
       "Assembly                                                         0         2   \n",
       "ChIP-Seq data analysis                                           1         2   \n",
       "Development in Galaxy                                            1        13   \n",
       "Epigenetics                                                      1         1   \n",
       "Galaxy Introduction                                              1         8   \n",
       "Metagenomics                                                     1         2   \n",
       "Proteomics                                                       1         7   \n",
       "Analysis of sequences                                            1         7   \n",
       "Train the trainers                                               1         6   \n",
       "Transcriptomics                                                  1         2   \n",
       "Tutorials using public Galaxy at usegalaxy.org                   2         7   \n",
       "Variant Analysis                                                 1         2   \n",
       "\n",
       "                                               Slide decks Hands-on tutorials  \\\n",
       "Galaxy Server administration                             2                  2   \n",
       "Assembly                                                 2                  2   \n",
       "ChIP-Seq data analysis                                   0                  2   \n",
       "Development in Galaxy                                   12                  4   \n",
       "Epigenetics                                              0                  1   \n",
       "Galaxy Introduction                                      2                  6   \n",
       "Metagenomics                                             0                  2   \n",
       "Proteomics                                               0                  7   \n",
       "Analysis of sequences                                    2                  7   \n",
       "Train the trainers                                       1                  5   \n",
       "Transcriptomics                                          0                  2   \n",
       "Tutorials using public Galaxy at usegalaxy.org           0                  7   \n",
       "Variant Analysis                                         0                  2   \n",
       "\n",
       "                                               Workflows Data on Zenodo  \\\n",
       "Galaxy Server administration                           0              0   \n",
       "Assembly                                               1              0   \n",
       "ChIP-Seq data analysis                                 1              0   \n",
       "Development in Galaxy                                  0              0   \n",
       "Epigenetics                                            1              0   \n",
       "Galaxy Introduction                                    0              0   \n",
       "Metagenomics                                           2              0   \n",
       "Proteomics                                             7              0   \n",
       "Analysis of sequences                                  7              0   \n",
       "Train the trainers                                     0              1   \n",
       "Transcriptomics                                        2              0   \n",
       "Tutorials using public Galaxy at usegalaxy.org         0              0   \n",
       "Variant Analysis                                       2              0   \n",
       "\n",
       "                                               Data on data library  \n",
       "Galaxy Server administration                                      0  \n",
       "Assembly                                                          2  \n",
       "ChIP-Seq data analysis                                            1  \n",
       "Development in Galaxy                                             0  \n",
       "Epigenetics                                                       1  \n",
       "Galaxy Introduction                                               8  \n",
       "Metagenomics                                                      2  \n",
       "Proteomics                                                        6  \n",
       "Analysis of sequences                                             7  \n",
       "Train the trainers                                                5  \n",
       "Transcriptomics                                                   2  \n",
       "Tutorials using public Galaxy at usegalaxy.org                    0  \n",
       "Variant Analysis                                                  2  "
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats_per_date['22.06.17']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Introduction slide decks</th>\n",
       "      <th>Tutorials</th>\n",
       "      <th>Slide decks</th>\n",
       "      <th>Hands-on tutorials</th>\n",
       "      <th>Workflows</th>\n",
       "      <th>Data on Zenodo</th>\n",
       "      <th>Data on data library</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Galaxy Server administration</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Assembly</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ChIP-Seq data analysis</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Contributing to the Galaxy Training Material</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Development in Galaxy</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Epigenetics</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Genome Annotation</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Introduction to Galaxy</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Metagenomics</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Proteomics</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Sequence analysis</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Statistics</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Transcriptomics</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Variant Analysis</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Introduction slide decks  \\\n",
       "Galaxy Server administration                                        1   \n",
       "Assembly                                                            0   \n",
       "ChIP-Seq data analysis                                              1   \n",
       "Contributing to the Galaxy Training Material                        1   \n",
       "Development in Galaxy                                               1   \n",
       "Epigenetics                                                         1   \n",
       "Genome Annotation                                                   0   \n",
       "Introduction to Galaxy                                              1   \n",
       "Metagenomics                                                        1   \n",
       "Proteomics                                                          1   \n",
       "Sequence analysis                                                   1   \n",
       "Statistics                                                          0   \n",
       "Transcriptomics                                                     1   \n",
       "Variant Analysis                                                    1   \n",
       "\n",
       "                                             Tutorials Slide decks  \\\n",
       "Galaxy Server administration                         9           7   \n",
       "Assembly                                             4           3   \n",
       "ChIP-Seq data analysis                               2           0   \n",
       "Contributing to the Galaxy Training Material        13           3   \n",
       "Development in Galaxy                               13          12   \n",
       "Epigenetics                                          2           0   \n",
       "Genome Annotation                                    2           1   \n",
       "Introduction to Galaxy                              13           2   \n",
       "Metagenomics                                         2           0   \n",
       "Proteomics                                           8           0   \n",
       "Sequence analysis                                    5           2   \n",
       "Statistics                                           1           0   \n",
       "Transcriptomics                                      5           1   \n",
       "Variant Analysis                                     7           0   \n",
       "\n",
       "                                             Hands-on tutorials Workflows  \\\n",
       "Galaxy Server administration                                  4         0   \n",
       "Assembly                                                      4         2   \n",
       "ChIP-Seq data analysis                                        2         2   \n",
       "Contributing to the Galaxy Training Material                 11         0   \n",
       "Development in Galaxy                                         4         0   \n",
       "Epigenetics                                                   2         2   \n",
       "Genome Annotation                                             2         2   \n",
       "Introduction to Galaxy                                       11         0   \n",
       "Metagenomics                                                  2         2   \n",
       "Proteomics                                                    8         8   \n",
       "Sequence analysis                                             5         5   \n",
       "Statistics                                                    1         1   \n",
       "Transcriptomics                                               5         4   \n",
       "Variant Analysis                                              7         4   \n",
       "\n",
       "                                             Data on Zenodo  \\\n",
       "Galaxy Server administration                              0   \n",
       "Assembly                                                  0   \n",
       "ChIP-Seq data analysis                                    0   \n",
       "Contributing to the Galaxy Training Material              2   \n",
       "Development in Galaxy                                     0   \n",
       "Epigenetics                                               0   \n",
       "Genome Annotation                                         0   \n",
       "Introduction to Galaxy                                    0   \n",
       "Metagenomics                                              0   \n",
       "Proteomics                                                0   \n",
       "Sequence analysis                                         0   \n",
       "Statistics                                                0   \n",
       "Transcriptomics                                           0   \n",
       "Variant Analysis                                          0   \n",
       "\n",
       "                                             Data on data library  \n",
       "Galaxy Server administration                                    0  \n",
       "Assembly                                                        3  \n",
       "ChIP-Seq data analysis                                          2  \n",
       "Contributing to the Galaxy Training Material                    0  \n",
       "Development in Galaxy                                           0  \n",
       "Epigenetics                                                     1  \n",
       "Genome Annotation                                               2  \n",
       "Introduction to Galaxy                                          8  \n",
       "Metagenomics                                                    2  \n",
       "Proteomics                                                      6  \n",
       "Sequence analysis                                               5  \n",
       "Statistics                                                      1  \n",
       "Transcriptomics                                                 4  \n",
       "Variant Analysis                                                4  "
      ]
     },
     "execution_count": 178,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats_per_date['22.06.18']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Introduction slide decks</th>\n",
       "      <th>Tutorials</th>\n",
       "      <th>Slide decks</th>\n",
       "      <th>Hands-on tutorials</th>\n",
       "      <th>Workflows</th>\n",
       "      <th>Data on Zenodo</th>\n",
       "      <th>Data on data library</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Galaxy Server administration</td>\n",
       "      <td>1</td>\n",
       "      <td>28</td>\n",
       "      <td>24</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Assembly</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ChIP-Seq data analysis</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Computational chemistry</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Contributing to the Galaxy Training Material</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Development in Galaxy</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Epigenetics</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Data Manipulation</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>User Interface and Features</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Genome Annotation</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Teaching and Hosting Galaxy training</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Introduction to Galaxy Analyses</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Metabolomics</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Metagenomics</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Proteomics</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Sequence analysis</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Statistics and machine learning</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Transcriptomics</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "      <td>16</td>\n",
       "      <td>14</td>\n",
       "      <td>16</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Variant Analysis</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Introduction slide decks  \\\n",
       "Galaxy Server administration                                        1   \n",
       "Assembly                                                            0   \n",
       "ChIP-Seq data analysis                                              1   \n",
       "Computational chemistry                                             0   \n",
       "Contributing to the Galaxy Training Material                        1   \n",
       "Development in Galaxy                                               1   \n",
       "Epigenetics                                                         1   \n",
       "Data Manipulation                                                   0   \n",
       "User Interface and Features                                         0   \n",
       "Genome Annotation                                                   1   \n",
       "Teaching and Hosting Galaxy training                                0   \n",
       "Introduction to Galaxy Analyses                                     1   \n",
       "Metabolomics                                                        1   \n",
       "Metagenomics                                                        1   \n",
       "Proteomics                                                          0   \n",
       "Sequence analysis                                                   0   \n",
       "Statistics and machine learning                                     0   \n",
       "Transcriptomics                                                     1   \n",
       "Variant Analysis                                                    1   \n",
       "\n",
       "                                             Tutorials Slide decks  \\\n",
       "Galaxy Server administration                        28          24   \n",
       "Assembly                                             4           3   \n",
       "ChIP-Seq data analysis                               1           1   \n",
       "Computational chemistry                              3           0   \n",
       "Contributing to the Galaxy Training Material        10           2   \n",
       "Development in Galaxy                               12          12   \n",
       "Epigenetics                                          2           0   \n",
       "Data Manipulation                                    6           1   \n",
       "User Interface and Features                          3           0   \n",
       "Genome Annotation                                    3           1   \n",
       "Teaching and Hosting Galaxy training                 5           1   \n",
       "Introduction to Galaxy Analyses                      7           1   \n",
       "Metabolomics                                         2           0   \n",
       "Metagenomics                                         3           0   \n",
       "Proteomics                                          13           0   \n",
       "Sequence analysis                                    5           2   \n",
       "Statistics and machine learning                      4           0   \n",
       "Transcriptomics                                     17           3   \n",
       "Variant Analysis                                     6           0   \n",
       "\n",
       "                                             Hands-on tutorials Workflows  \\\n",
       "Galaxy Server administration                                 13         0   \n",
       "Assembly                                                      4         4   \n",
       "ChIP-Seq data analysis                                        1         1   \n",
       "Computational chemistry                                       3         2   \n",
       "Contributing to the Galaxy Training Material                  9         1   \n",
       "Development in Galaxy                                         3         0   \n",
       "Epigenetics                                                   2         2   \n",
       "Data Manipulation                                             5         0   \n",
       "User Interface and Features                                   3         0   \n",
       "Genome Annotation                                             3         2   \n",
       "Teaching and Hosting Galaxy training                          4         0   \n",
       "Introduction to Galaxy Analyses                               6         2   \n",
       "Metabolomics                                                  2         2   \n",
       "Metagenomics                                                  3         3   \n",
       "Proteomics                                                   13        12   \n",
       "Sequence analysis                                             5         5   \n",
       "Statistics and machine learning                               4         3   \n",
       "Transcriptomics                                              16        14   \n",
       "Variant Analysis                                              6         4   \n",
       "\n",
       "                                             Data on Zenodo  \\\n",
       "Galaxy Server administration                              7   \n",
       "Assembly                                                  4   \n",
       "ChIP-Seq data analysis                                    1   \n",
       "Computational chemistry                                   3   \n",
       "Contributing to the Galaxy Training Material              3   \n",
       "Development in Galaxy                                     0   \n",
       "Epigenetics                                               2   \n",
       "Data Manipulation                                         5   \n",
       "User Interface and Features                               3   \n",
       "Genome Annotation                                         3   \n",
       "Teaching and Hosting Galaxy training                      0   \n",
       "Introduction to Galaxy Analyses                           6   \n",
       "Metabolomics                                              2   \n",
       "Metagenomics                                              3   \n",
       "Proteomics                                               13   \n",
       "Sequence analysis                                         5   \n",
       "Statistics and machine learning                           4   \n",
       "Transcriptomics                                          16   \n",
       "Variant Analysis                                          6   \n",
       "\n",
       "                                             Data on data library  \n",
       "Galaxy Server administration                                    0  \n",
       "Assembly                                                        4  \n",
       "ChIP-Seq data analysis                                          1  \n",
       "Computational chemistry                                         0  \n",
       "Contributing to the Galaxy Training Material                    0  \n",
       "Development in Galaxy                                           0  \n",
       "Epigenetics                                                     2  \n",
       "Data Manipulation                                               0  \n",
       "User Interface and Features                                     1  \n",
       "Genome Annotation                                               3  \n",
       "Teaching and Hosting Galaxy training                            0  \n",
       "Introduction to Galaxy Analyses                                 2  \n",
       "Metabolomics                                                    1  \n",
       "Metagenomics                                                    2  \n",
       "Proteomics                                                      9  \n",
       "Sequence analysis                                               5  \n",
       "Statistics and machine learning                                 4  \n",
       "Transcriptomics                                                13  \n",
       "Variant Analysis                                                6  "
      ]
     },
     "execution_count": 179,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats_per_date['22.06.19']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Introduction slide decks</th>\n",
       "      <th>Tutorials</th>\n",
       "      <th>Slide decks</th>\n",
       "      <th>Hands-on tutorials</th>\n",
       "      <th>Workflows</th>\n",
       "      <th>Data on Zenodo</th>\n",
       "      <th>Data on data library</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Galaxy Server administration</td>\n",
       "      <td>1</td>\n",
       "      <td>31</td>\n",
       "      <td>24</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Assembly</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Computational chemistry</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Contributing to the Galaxy Training Material</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Development in Galaxy</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Ecology</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Epigenetics</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Data Manipulation</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>User Interface and Features</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Genome Annotation</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Imaging</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Teaching and Hosting Galaxy training</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Introduction to Galaxy Analyses</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Metabolomics</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Metagenomics</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Proteomics</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "      <td>14</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Sequence analysis</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Statistics and machine learning</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Transcriptomics</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>3</td>\n",
       "      <td>18</td>\n",
       "      <td>16</td>\n",
       "      <td>18</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Variant Analysis</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Introduction slide decks  \\\n",
       "Galaxy Server administration                                        1   \n",
       "Assembly                                                            0   \n",
       "Computational chemistry                                             0   \n",
       "Contributing to the Galaxy Training Material                        1   \n",
       "Development in Galaxy                                               1   \n",
       "Ecology                                                             0   \n",
       "Epigenetics                                                         2   \n",
       "Data Manipulation                                                   0   \n",
       "User Interface and Features                                         0   \n",
       "Genome Annotation                                                   1   \n",
       "Imaging                                                             0   \n",
       "Teaching and Hosting Galaxy training                                0   \n",
       "Introduction to Galaxy Analyses                                     1   \n",
       "Metabolomics                                                        1   \n",
       "Metagenomics                                                        1   \n",
       "Proteomics                                                          0   \n",
       "Sequence analysis                                                   0   \n",
       "Statistics and machine learning                                     0   \n",
       "Transcriptomics                                                     1   \n",
       "Variant Analysis                                                    1   \n",
       "\n",
       "                                             Tutorials Slide decks  \\\n",
       "Galaxy Server administration                        31          24   \n",
       "Assembly                                             4           3   \n",
       "Computational chemistry                              4           0   \n",
       "Contributing to the Galaxy Training Material        10           2   \n",
       "Development in Galaxy                               13          12   \n",
       "Ecology                                              5           0   \n",
       "Epigenetics                                          5           2   \n",
       "Data Manipulation                                    7           1   \n",
       "User Interface and Features                          4           0   \n",
       "Genome Annotation                                    4           2   \n",
       "Imaging                                              2           0   \n",
       "Teaching and Hosting Galaxy training                 5           1   \n",
       "Introduction to Galaxy Analyses                      7           1   \n",
       "Metabolomics                                         3           0   \n",
       "Metagenomics                                         4           0   \n",
       "Proteomics                                          14           0   \n",
       "Sequence analysis                                    2           2   \n",
       "Statistics and machine learning                      4           0   \n",
       "Transcriptomics                                     19           3   \n",
       "Variant Analysis                                     6           0   \n",
       "\n",
       "                                             Hands-on tutorials Workflows  \\\n",
       "Galaxy Server administration                                 16         0   \n",
       "Assembly                                                      4         4   \n",
       "Computational chemistry                                       4         3   \n",
       "Contributing to the Galaxy Training Material                  9         1   \n",
       "Development in Galaxy                                         4         0   \n",
       "Ecology                                                       5         4   \n",
       "Epigenetics                                                   5         5   \n",
       "Data Manipulation                                             6         0   \n",
       "User Interface and Features                                   4         1   \n",
       "Genome Annotation                                             4         2   \n",
       "Imaging                                                       2         2   \n",
       "Teaching and Hosting Galaxy training                          4         0   \n",
       "Introduction to Galaxy Analyses                               6         2   \n",
       "Metabolomics                                                  3         3   \n",
       "Metagenomics                                                  4         4   \n",
       "Proteomics                                                   14        13   \n",
       "Sequence analysis                                             2         2   \n",
       "Statistics and machine learning                               4         3   \n",
       "Transcriptomics                                              18        16   \n",
       "Variant Analysis                                              6         4   \n",
       "\n",
       "                                             Data on Zenodo  \\\n",
       "Galaxy Server administration                              6   \n",
       "Assembly                                                  4   \n",
       "Computational chemistry                                   4   \n",
       "Contributing to the Galaxy Training Material              3   \n",
       "Development in Galaxy                                     0   \n",
       "Ecology                                                   4   \n",
       "Epigenetics                                               5   \n",
       "Data Manipulation                                         6   \n",
       "User Interface and Features                               4   \n",
       "Genome Annotation                                         4   \n",
       "Imaging                                                   2   \n",
       "Teaching and Hosting Galaxy training                      0   \n",
       "Introduction to Galaxy Analyses                           6   \n",
       "Metabolomics                                              3   \n",
       "Metagenomics                                              4   \n",
       "Proteomics                                               14   \n",
       "Sequence analysis                                         2   \n",
       "Statistics and machine learning                           4   \n",
       "Transcriptomics                                          18   \n",
       "Variant Analysis                                          6   \n",
       "\n",
       "                                             Data on data library  \n",
       "Galaxy Server administration                                    0  \n",
       "Assembly                                                        4  \n",
       "Computational chemistry                                         0  \n",
       "Contributing to the Galaxy Training Material                    0  \n",
       "Development in Galaxy                                           0  \n",
       "Ecology                                                         4  \n",
       "Epigenetics                                                     4  \n",
       "Data Manipulation                                               0  \n",
       "User Interface and Features                                     1  \n",
       "Genome Annotation                                               4  \n",
       "Imaging                                                         2  \n",
       "Teaching and Hosting Galaxy training                            0  \n",
       "Introduction to Galaxy Analyses                                 2  \n",
       "Metabolomics                                                    2  \n",
       "Metagenomics                                                    3  \n",
       "Proteomics                                                      9  \n",
       "Sequence analysis                                               2  \n",
       "Statistics and machine learning                                 4  \n",
       "Transcriptomics                                                13  \n",
       "Variant Analysis                                                6  "
      ]
     },
     "execution_count": 180,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats_per_date['today']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Global informations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>22.06.17</th>\n",
       "      <th>22.06.18</th>\n",
       "      <th>22.06.19</th>\n",
       "      <th>today</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Topics</td>\n",
       "      <td>13</td>\n",
       "      <td>14</td>\n",
       "      <td>19</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Introduction slide decks</td>\n",
       "      <td>13</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Tutorials</td>\n",
       "      <td>63</td>\n",
       "      <td>86</td>\n",
       "      <td>134</td>\n",
       "      <td>153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Slide decks</td>\n",
       "      <td>21</td>\n",
       "      <td>31</td>\n",
       "      <td>51</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Hands-on tutorials</td>\n",
       "      <td>49</td>\n",
       "      <td>68</td>\n",
       "      <td>105</td>\n",
       "      <td>124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Workflows</td>\n",
       "      <td>23</td>\n",
       "      <td>32</td>\n",
       "      <td>57</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Data on Zenodo</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>86</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Data on data library</td>\n",
       "      <td>36</td>\n",
       "      <td>38</td>\n",
       "      <td>53</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          22.06.17  22.06.18  22.06.19  today\n",
       "Topics                          13        14        19     20\n",
       "Introduction slide decks        13        11        11     11\n",
       "Tutorials                       63        86       134    153\n",
       "Slide decks                     21        31        51     53\n",
       "Hands-on tutorials              49        68       105    124\n",
       "Workflows                       23        32        57     69\n",
       "Data on Zenodo                   1         2        86     99\n",
       "Data on data library            36        38        53     60"
      ]
     },
     "execution_count": 181,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "global_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAagAAAEYCAYAAAAJeGK1AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nO3deZxU1Z3+8c8DDXYU0LCIklYxqAEBIdKImcyovzFojI4LZNGYUYKIyzAxo6JMxgUNBpcxmp/gglEkxgUxEBWXRIS4JDGKI1uLS4gaQB0bAi2tIoLf+ePeJkXTSEtV07eqn/frVS/7nrudart46px77j2KCMzMzLKmVXNXwMzMrCEOKDMzyyQHlJmZZZIDyszMMskBZWZmmVTW3BVoSOfOnaN79+7NXQ0zM9sOXnjhhRUR0aV+eSYDqnv37sydO7e5q2FmZtuBpDcbKncXn5mZZZIDyszMMskBZWZmmZTJa1AN+fjjj1m2bBlr165t7qoUpfLycioqKmjTpk1zV8XMrFGKJqCWLVtG+/bt6d69O5KauzpFJSJYuXIly5YtY++9927u6piZNUrRdPGtXbuWTp06OZy2gSQ6derk1qeZFZWiCSjA4ZQH/+7MrNgUVUCZmVnL0ahrUJJGAcOAvsA9ETEsZ92OwH8D3wbaAPMj4pB0nYArgRHp5rcBF0YBJqHqPubhfA+xiTeuPHqr2yxdupRTTjmFd955h1atWjFy5EjOOeccRo8ezUMPPUTbtm3p0aMHkydPZpdddtls/8cee4xzzjmHDRs2MGLECMaMGbNx3Q033MCECRMoKyvj6KOP5uqrr95s/+HDhzNz5kx23XVXFi1atLH8O9/5Dq+88goAq1evZpdddmHevHnb8mswM8uMxg6SeAsYBxwJfK7euknpcXoBfwP656wbCRwP9AMCeBz4C3Dztle5+ZSVlXHttddy4IEHsmbNGgYMGMDgwYMZPHgw48ePp6ysjAsvvJDx48dz1VVXbbLvhg0b+Ld/+zcef/xxKioqGDhwIMceeyz7778/c+bM4YEHHmDBggXssMMOvPvuuw2ef9iwYYwaNYpTTjllk/KpU6du/Pm8885j5513LvybN7NttrhnryY9fq+XFzfp8ZtLo7r4ImJ6RPwaWJlbLulLwLHAyIiojogNEfFCzianAtdGxLKIWA5cS9ISK0q77747Bx54IADt27enV69eLF++nCOOOIKysiTrDz74YJYtW7bZvs899xz77LMPX/ziF2nbti0nnngiDzzwAAA33XQTY8aMYYcddgBg1113bfD8hxxyCB07dtxi/SKC++67j5NOOimv92lmlgX5XoMaBLwJXCZphaSFkobmrO8NzM9Znp+WbUbSSElzJc2trq7Os1pN74033uDFF19k0KBBm5TffvvtHHXUUZttv3z5cvbYY4+NyxUVFSxfvhyAV199laeffppBgwZx6KGH8vzzz29TnZ5++mm6du3Kvvvuu037m5llSb4BVQH0AWqAbsAoYIqkuvZsu3RdnRqgnRoYUhYRkyKiMiIqu3TZ7KG2mVJbW8vQoUO5/vrr6dChw8byK664grKyMk4++eTN9mnoslvdr2H9+vWsWrWKZ599lmuuuYZvf/vbDW6/Nffcc49bT2ZWMvK9UfdD4GNgXESsB56UNAc4AlgM1AIdcrbvANQWYpBEc/n4448ZOnQoJ598MkOGDNlYPmXKFGbOnMkTTzzR4JDuiooKli5dunF52bJldOvWbeO6IUOGIImDDjqIVq1asWLFCj5LUK9fv57p06fzwgsvbH1jM7MikG8LasFW1leRDJCo0y8tK0oRwWmnnUavXr0499xzN5Y/9thjXHXVVTz44IPsuOOODe47cOBAXnvtNV5//XXWrVvHvffey7HHHgvA8ccfz+zZs4Gku2/dunV07tz5M9Vt1qxZ9OzZk4qKim18d2Zm2dLYYeZl6batgdaSyoH1wFPAX4H/lDSe5JrUYcDodNdfAOdKeoRkFN95wA2FqHhjhoUX2u9//3vuvPNO+vbtS//+yWDFn/zkJ/zgBz/go48+YvDgwUAyUOLmm2/mrbfeYsSIETzyyCOUlZUxYcIEjjzySDZs2MDw4cPp3Tu5HDd8+HCGDx9Onz59aNu2LVOmTEHSJvsDnHTSSfzud79jxYoVVFRUcNlll3HaaacBcO+997p7z8xKihrT2yZpLHBpveLLImKspN7Az4EDSAZM/FdEzEj3E3AVf78P6uc04j6oysrKqD9h4eLFi+nVq2mHapY6/w7NmoeHmX86SS9ERGX98ka1oCJiLDB2C+uqgK9sYV0AF6QvMzOzRvOjjszMLJMcUGZmlkkOKDMzyyQHlJmZZZIDyszMMqlopnzfzNgCP7F7bM1WN8nqdBvz5s3jzDPPZO3atZSVlXHjjTdy0EEHbeMvwswsG9yC+gzqpttYvHgxzz77LBMnTuSll15i8ODBLFq0iAULFrDffvsxfvz4zfatm27j0Ucf5aWXXuKee+7hpZdeAthkuo2qqirOP//8Bs8/bNgwHnvssc3KL7jgAi699FLmzZvH5ZdfzgUXeFS/mRU/B9RnkNXpNiTx3nvvAVBTU7PxGX9mZsXMAbWNsjTdxvXXX8/o0aPZY489OP/88xtswZmZFRsH1DbI2nQbN910E9dddx1Lly7luuuu2/h8PjOzYuaA+oy2Nt3GXXfdVbDpNhprypQpG+vyrW99i+eee25b356ZWWY4oD6DrE630a1bN5588kkAZs+e7Rl1zawkFPEw860PCy+0rE63ceutt3LOOeewfv16ysvLmTRp0nb/3ZiZFVqjptvY3jzdRtPw79CseXi6jU+X13QbZmalrO+Uvk16/Pua9Oilq1HXoCSNkjRX0keS7tjCNpdKCklfyymTpKskrUxfV6uhEQRmZmb1NLYF9RYwDjgS+Fz9lZJ6AN8E3q63aiRwPNCPZMr3x4G/ADdvY33NzKyFaFQLKiKmR8SvgZVb2GQCcCGwrl75qcC1EbEsIpYD1wLDtrGuZmbWguQ9zFzSt4B1EfFIA6t7A/NzluenZQ0dZ2TajTi3uro632qZmVmRyyugJLUDfgL8cAubtANyx4PXAO0aug4VEZMiojIiKrt06ZJPtczMrATkO4rvMuDOiHh9C+trgQ45yx2A2ijA2PZCj7pZeOrCrW6T1ek25s+fz5lnnkltbS3du3fnrrvu2uQRTGZmxSjfLr7DgR9IekfSO8AewH2SLkzXV5EMkKjTLy0rSlmdbmPEiBFceeWVLFy4kBNOOIFrrrmmsG/czKwZNHaYeZmkcqA10FpSuaQykoDqA/RPX28BZwAT011/AZwr6QuSugHnAXcU9i1sP1mdbuOVV17hkEMOAWDw4MH86le/yv/Nmpk1s8a2oC4CPgTGAN9Lf74oIlZGxDt1L2ADsCoiatP9bgEeAhYCi4CH07Kil6XpNvr06cODDz4IwLRp0zZ5KK2ZWbFq7DDzsRGheq+xDWzXPSJm5SxHRFwQER3T1wWFuP7U3LI23cbtt9/OxIkTGTBgAGvWrKFt27bb8K7MzLLFjzr6jLY23cYTTzxRsOk2GjuasWfPnvz2t78FktbYww8/nM9bNDPLBE+38RlkdbqNd999F4BPPvmEcePGceaZZ27rWzQzy4yibUE1Zlh4oWV1uo177rmHiROTcSlDhgzh+9///nb/3ZiZFZqn22hB/Ds0a1iTP818/PomPX6pTrfhLj4zM8skB5SZmWWSA8rMzDLJAWVmZpnkgDIzs0xyQJmZWSYV7X1Qi3sWdrj01oZprl69mrvvvpuzzz670cccNmwYxxxzDN/85jfzrZ6ZWYvjFlQjrV69mhtvvLG5q2Fm1mI4oBppzJgxLFmyhP79+zN69GhGjx5Nnz596Nu3L1OnTgWSRyGNGjWK/fffn6OPPnrjI4gALr/8cgYOHEifPn0YOXIkEcGSJUs2Tt8B8NprrzFgwIDt/t7MzLLIAdVIV155JT169GDevHkcfPDBzJs3j/nz5zNr1ixGjx7N22+/zYwZM3jllVdYuHAht956K3/4wx827j9q1Cief/55Fi1axIcffsjMmTPp0aMHO++8M/PmzQNg8uTJDBs2rJneoZlZtjigtsEzzzzDSSedROvWrenatevGOZyeeuqpjeXdunXjn//5nzfuM2fOHAYNGkTfvn2ZPXs2VVXJxMIjRoxg8uTJbNiwgalTp/Ld7363ud6WmVmmNHZG3VGS5kr6SNIdOeUHS3pc0t8kVUuaJmn3nPWSdJWklenrajU0F0WR+bTnFzb09tauXcvZZ5/N/fffz8KFCzn99NNZu3YtAEOHDuXRRx9l5syZDBgwgE6dOjVZvc3MikljW1BvAeOA2+uVfx6YBHQH9gLWAJNz1o8Ejgf6AQcAx5BMCV902rdvz5o1a4Bk6vWpU6eyYcMGqqureeqppzjooIM45JBDuPfee9mwYQNvv/02c+bMAdgYRp07d6a2tpb7779/43HLy8s58sgjOeuss/wUcjOzHI0aZh4R0wEkVQIVOeWP5m4naQLwZE7RqcC1EbEsXX8tcDpwc37V3v5P7+3UqRNf/epX6dOnD0cddRQHHHAA/fr1QxJXX301u+22GyeccAKzZ8+mb9++7Lfffhx66KEA7LLLLpx++un07duX7t27M3DgwE2OffLJJzN9+nSOOOKI7fqezMyyrND3QR0CVOUs9wbm5yzPT8s2I2kkSYuLPffcs8DVKoy77757k+Vrrrlmk2VJTJgwocF9x40bx7hx4xpc98wzzzB8+HBat25dmIqamZWAggWUpAOAS4DjcorbATU5yzVAO0mKehdyImISSXchlZWV2ZukqomccMIJLFmyZOOMumZmlihIQEnaB3gUOCcins5ZVQt0yFnuANTWD6eWbMaMGc1dBTOzTMp7mLmkvYBZwI8j4s56q6tIBkjU6cemXYCfiXNt2/l3Z2bFprHDzMsklQOtgdaSytOyLwCzgYkR0dDAh18A50r6gqRuwHnAHdtS0fLyclauXOl/aLdBRLBy5UrKy8ubuypmZo3W2C6+i4BLc5a/B1wGBPBF4FJJG9dHRLv0x1vS9QvT5Z+nZZ9ZRUUFy5Yto7q6elt2b/HKy8upqKjY+oZmZhnR2GHmY4GxW1h92afsF8AF6Ssvbdq0Ye+99873MGZmViT8qCMzM8skB5SZmWWSA8rMzDLJAWVmZpnkgDIzs0xyQJmZWSY5oMzMLJMcUGZmlkkOKDMzyyQHlJmZZZIDyszMMskBZWZmmeSAMjOzTHJAmZlZJjmgzMwskxo7o+4oSXMlfSTpjnrrDpf0sqQPJM1Jp4CvWydJV0lamb6ulqQCvwczMytBjW1BvQWMA27PLZTUGZgOXAx0BOYCU3M2GQkcD/QDDgCOAc7Ir8pmZtYSNCqgImJ6RPwaWFlv1RCgKiKmRcRakll3+0nqma4/Fbg2IpZFxHLgWmBYQWpuZmYlLd9rUL2B+XULEfE+sCQt32x9+nNvGiBpZNqNOLe6ujrPapmZWbHLN6DaATX1ymqA9ltYXwO0a+g6VERMiojKiKjs0qVLntUyM7Nil29A1QId6pV1ANZsYX0HoDYiIs/zmplZics3oKpIBkAAIGknoEdavtn69OcqzMzMtqKxw8zLJJUDrYHWksollQEzgD6ShqbrLwEWRMTL6a6/AM6V9AVJ3YDzgDsK/i7MzKzkNLYFdRHwITAG+F7680URUQ0MBa4AVgGDgBNz9rsFeAhYCCwCHk7LzMzMPlVZYzaKiLEkQ8gbWjcL6LmFdQFckL7MzMwazY86MjOzTHJAmZlZJjmgzMwskxxQZmaWSQ4oMzPLJAeUmZllkgPKzMwyyQFlZmaZ5IAyM7NMckCZmVkmOaDMzCyTGvUsPjOzZjd256Y79t57Nt2xbZu5BWVmZpnkgDIzs0xyQJmZWSYVJKAkdZf0iKRVkt6RNCGdcRdJh0t6WdIHkuZI2qsQ5zQzs9JWqBbUjcC7wO5Af+BQ4GxJnYHpwMVAR2AuMLVA5zQzsxJWqIDaG7gvItZGxDvAY0BvYAhQFRHTImItyay8/SQ1OAOvmZlZnUIF1M+AEyXtKOkLwFH8PaTm120UEe8DS9LyTUgaKWmupLnV1dUFqpaZmRWrQgXUkySh8x6wjKQr79dAO6Cm3rY1QPv6B4iISRFRGRGVXbp0KVC1zMysWOUdUJJaAb8huda0E9AZ+DxwFVALdKi3SwdgTb7nNTOz0laIFlRHYA9gQkR8FBErgcnAN4AqoF/dhpJ2Anqk5WZmZluUd0BFxArgdeAsSWWSdgFOJbn2NAPoI2mopHLgEmBBRLyc73nNzKy0Feoa1BDg60A18GdgPfAfEVENDAWuAFYBg4ATC3ROMzMrYQV5WGxEzAMO28K6WYCHlZuZ2WfiRx2ZmVkmOaDMzCyTHFBmZpZJDigzM8skB5SZmWWSA8rMzDLJAWVmZpnkgDIzs0xyQJmZWSY5oMzMLJMcUGZmlkkOKDMzyyQHlJmZZZIDyszMMskBZWZmmVSQ+aAAJJ0IXArsCbwDDIuIpyUdDkxMy/+Ulr9ZqPOaWTZ0H/Nwkx7/jfImPbxlUEFaUJIGA1cB3wfaA4cAf5HUGZgOXAx0BOYCUwtxTjMzK22FakFdBlweEc+my8sBJI0EqiJiWro8FlghqWdEvFygc5uZWQnKuwUlqTVQCXSR9GdJyyRNkPQ5oDcwv27biHgfWJKW1z/OSElzJc2trq7Ot1pmZlbkCtHF1xVoA3wT+CegP/Bl4CKgHVBTb/sakm7ATUTEpIiojIjKLl26FKBaZmZWzAoRUB+m/70hIt6OiBXAT4FvALVAh3rbdwDWFOC8ZmZWwvIOqIhYBSwDooHVVUC/ugVJOwE90nIzM7MtKtR9UJOBf5e0q6TPAz8EZgIzgD6ShkoqBy4BFniAhJmZbU2hAurHwPPAq8Bi4EXgioioBoYCVwCrgEHAiQU6p5mZlbCCDDOPiI+Bs9NX/XWzgJ6FOI+ZmbUcftSRmZllkgPKzMwyyQFlZmaZ5IAyM7NMckCZmVkmOaDMzCyTHFBmZpZJDigzM8skB5SZmWWSA8rMzDLJAWVmZpnkgDIzs0xyQJmZWSY5oMzMLJMcUGZmlkkFCyhJ+0paK+mXOWWHS3pZ0geS5kjaq1DnMzOz0lbIFtREkll1AZDUGZgOXAx0BOYCUwt4PjMzK2EFCShJJwKrgSdyiocAVRExLSLWAmOBfpI8u66ZmW1V3gElqQNwOXBevVW9gfl1CxHxPrAkLW/oOCMlzZU0t7q6Ot9qmZlZkStEC+rHwG0RsbReeTugpl5ZDdC+oYNExKSIqIyIyi5duhSgWmZmVszK8tlZUn/ga8CXG1hdC3SoV9YBWJPPOc3MrGXIK6CAw4DuwF8lQdJqai1pf+Bm4NS6DSXtBPQAqvI8p5mZtQD5dvFNIgmd/unrZuBh4EhgBtBH0lBJ5cAlwIKIeDnPc5qZWQuQVwsqIj4APqhbllQLrI2I6nR5KDAB+CXwJ+DEfM5nZmYtR75dfJuIiLH1lmcBHlZuZmafmR91ZGZmmeSAMjOzTHJAmZlZJjmgzMwskxxQZmaWSQ4oMzPLJAeUmZllkgPKzMwyyQFlZmaZ5IAyM7NMckCZmVkmOaDMzCyTHFBmZpZJDigzM8skB5SZmWVS3gElaQdJt0l6U9IaSS9KOipn/eGSXpb0gaQ5kvbK95xmZlb6CtGCKgOWAocCOwMXA/dJ6i6pMzA9LesIzAWmFuCcZmZW4vKeUTci3gfG5hTNlPQ6MADoBFRFxDQASWOBFZJ6RsTL+Z7bzMxKV0GnfAeQ1BXYD6gCzgLm162LiPclLQF6Ay/X228kMBJgzz33LHS1LCMW9+zVpMfv9fLiJj2+mW0/BR0kIakNcBcwJW0htQNq6m1WA7Svv29ETIqIyoio7NKlSyGrZWZmRahgLShJrYA7gXXAqLS4FuhQb9MOwJpCndcKq++Uvk16/Pua9OhmVkoK0oKSJOA2oCswNCI+TldVAf1yttsJ6JGWm5mZbVGhuvhuAnoB/xIRH+aUzwD6SBoqqRy4BFjgARJmZrY1hbgPai/gDKA/8I6k2vR1ckRUA0OBK4BVwCDgxHzPaWZmpa8Qw8zfBPQp62cBPfM9j5mZtSx+1JGZmWVSwe+DsiY2duemPf7evgfNzLLBAdUEuo95uMmO/UZ5kx3azCxT3MVnZmaZ5IAyM7NMckCZmVkmOaDMzCyTHFBmZpZJDigzM8skB5SZmWWSA8rMzDLJAWVmZpnkgDIzs0xyQJmZWSY5oMzMLJOaPKAkdZQ0Q9L7kt6U9N2mPqeZmRW/7fE084nAOqAryay7D0uaHxFV2+HcZmZWpJq0BSVpJ5Ip3y+OiNqIeAZ4EPjXpjyvmZkVv6ZuQe0HbIiIV3PK5gOH1t9Q0khgZLpYK+mVJq5bUVKTn2FRZ2BFUx19/6Y6cB01/W/ImkfT/p/1330z26uhwqYOqHZATb2yGqB9/Q0jYhIwqYnrY1shaW5EVDZ3Pcy2J//dZ1NTD5KoBTrUK+sArGni85qZWZFr6oB6FSiTtG9OWT/AAyTMzOxTNWlARcT7wHTgckk7SfoqcBxwZ1Oe1/LiblZrifx3n0GKiKY9gdQRuB0YDKwExkTE3U16UjMzK3pNHlBmZmbbwo86MjOzTHJAmZlZJjmgzMy2gVT8d8dmnQPKzOwzklQW6QV8SeXNXZ9S5YCyRpPUOv1vB0mbPQ3ErCWQ1Doi1ktqJek24ARJbZq7XqXIAWWNkn4oN0jqC0wFviJph+aul9n2ln4OWgEvkDwZ51FgQ/PWqjQ5oKxR0g9lH+B3wB8i4rcR8VEzV8usyUnaT9K36xX/CFgcEd+KiNXA4ZKGSGrwoae2bbbHfFBWAtLuvbOBH0fE9ek3yO+T/A0tiIg/NmsFzZrOQOBOSW0i4q50cEQ58L6k7sBVwJdInobeR1LfiKhuttqWEAeUbVFOt17df8uAr0t6DrgZeBdYCwyTdFxEvNusFTZrAmkotQV+IalVRNwp6ZckvQk7kvw7+lVgF2BCWmYF4ICyBuWE0v7A8ZJ+DlwK/Bo4DXgoIv5LUiVwIeDuPis5khSJyWnLaUo6gm+ypF7Ax8CH6WflNJKpnz5o1kqXEAeUbSLnA1l3zelJ4KdARMTbkv6RZOHjdJd/J/nm6ClUrKSkraVP6pYj4va0q/s2SZ9ExJR0u/6Svg98D/iau/cKxwFlAEj6IvB6zr0dOwDjgPER8d9120XEx0oMAH4C7A4MiIhP6n+gzYpV2kpan7aaepBMtLo6Im5Ny25PQ+pOYBXJ5KyHRsSiZqx2yXFAGZK+ApwJnMffp73egSR8FqbbtAY+SQNsN2A18DRwZfpBLouI9du98mYFlnufE/Abkr/3FcCfJF0WEZPSkLpV0ufS5TP89194HmZuAC8Bl0XEivQmXEXEe8AS4MD0Q7ghIiIdKHEcUBMR49IPcmt/OK2YpYMggI23VAj4PVANfA24BfgGcL2knSLiFpJrr+Ml7Yzvg2oSDqgWLu2Wq4mIv0jaDbgeODVd/RTw/4DjJHVLy+4ATgT+VneMiPCH04pW2r19t6TP5xR/HfhLRHw3Iv4XOJpkdF534Jo0pH4G9Eg/P563qAm4i6+Fq3fNaC3Jl5ZvSFoVETdL6gqcDFwnaS5Jt99X0mtO8gfTSsBfgfsiYlVO2f8A/wsgaQLJ6Lz+wDXAd4DWks4E3tvOdW1RPGFhC5YzlLwN0CYiPpDUARhLcmH4toh4UNKuQCXJxeDn6u6JcreeFbu6z0D6s4D/D9wcEVVp2W7AXcD5EfGipEtJhpH/MiLebq56txQOqBaqrvUjqR9wCcmF4MeA6cCfSUbofRG4G/hVbksr90NtVsxyb0JPr6f+kWQA0A+BV4EKks/EL4BOwHDgsIj4S7NVugXxNagWKL3uFJI6A9OAF4E7gQEkYTUIuBh4k2R03+G5+zucrBSkn4O6ARHvpF12/wS0IWlJ7RsRS4FZJNekjgGOdzhtP25BtVCS9iS5z+ntiLgwLduH5JtjWUScKakTcArwM9/fZKUoDadjgW9ExBlpWRnJ8PIATouINyW1A4iI2marbAvkFlQLlH4o+5B8Kzyirjwi/kzS336qpAERsTIirqu7CbeZqmvWlG4EppDc90fOtdUjSYaOT5O0T0TUOpy2P/+j00LkBkw68u63wBlAd0lX5Kz7I/An6t3X4RaUlYIGvmj9J/As0DsNovXp9dn1JF16bwPrtnc9LeEuvhYg50LwnsA/ArUk93gskjSU5N6n2cADJDcjHgL08rUmKyW5I0/T0aodI+INJbNDP0ZyU+656T2BvoUiAxxQJaihD1f64NffAHOBjkBX4JKIuFfSN0mmCXiX5HrTbek+fraelYS6v+WcxxetBw4EbiWZz6muV2E58KOIeK3ZKmsbuYuvxKStpZD0OUkHpGU7AlcCV0fEccBQ4GrgRklHRMT9wFkkU2Z4RlArOTlftB4F3gFGAsNIegyuTK8vHQUcAFySDpSwZub/CSUkpyuvA1AFXEYy2+0H6TfHlQAR8a6k20huxj2B5Jvj4+lhxqaPcTnPrScrRrk9CPV+bg+0Bv4jIlYASyUtBx6WtCAibpE0kKTrzzehZ4BbUCUiJ5zakwxymBkRP0/X7URy9/t+6XLdh/avJN19dcNnHwHGA/+Q3iNlVnRyAukLOT9/haRbb0/gX3O2XQBMBg5OuwHfi4g3tn+trSEOqBJQL5yeAtZHxFk5694HJgJjJP0Q2CPdtRKoSYedExEfATOAwek3TLOiJOlHwF2SekqqIpmr6UOSa05flvT1nM13JJnryT0GGeNBEiUi7dZ7HlhG8niWW4HJEbEy57FG3wB+RvIk8o9J7v34h0gmIfSACCspkp4G+gLPRMQxaVkvkpvRdycZPr6Y5PrrYeHJBjPH16BKgJLJBN8ApkXEGZKOAm4ieeLyLRGxOg2pRyS9AuwK7ATM8YNfrZTUe05kFbAP8FF6i8U7EbFY0k+AnsC3SEbvOZwyyi2oEiFpv4h4NWf5X4AbSILqlohYvYX9/OBXKylpl/WXI/rrBz8AAAIwSURBVOJ/0uWnSaZsPxf4czrcvGsk8zxZhvkaVImoC6d06gwi4iHg30m6L85QMutnQ/s5nKzUfAf4o6ST0uWjgc8D/w3sk16HvaHu+XqWXW5BlThJx5Bcd7qHv9/vYVbUPu1JD5I6kswK/QPgP9Ob0XcCHiK59tofOKquhWXZ5WtQJS4iZkr6HMm3yvebuz5m+ap3b9N+EfFq2q1XGRHPR8TfJN0BCLhS0icRcV96bXZfoNrde8XBLagSV/+bpp8xZsWsXjg9CrwYET9Kh5UPAiZGxG/T9bsAo0nuezovIqY1V71t2/gaVImrH0YOJytW9cLpVyRPfPhRuvohktsnTqq7xykdGDQfWAtcJKl93T1/VhwcUGaWefXC6Q7gMNKZniW1iYiFJA99bQV8W9Jx6a47A3eTDCVf4y9oxcVdfGZWNCT9mmSQQwA/Be7JfeqJpJ7AOSRPSVkH9CIJpwXNUF3LkwPKzIqCpN8BO0bEQZIOI5kJ90bq3ecnaTfgS0Bv4HFPnVG8HFBmVhQkHZg7NLyxN6Nb8fIwczMrCjlPhmgbEesi4qF0zMMNafnNEVHTnHW0wnJAmVlRiYh1OT/XhdRPgR0lXRsR7zVb5aygPIrPzIpa+livMcDxQJtmro4VkK9BmVlRy5lOpp0f5VVaHFBmVhL8lJTS44AyM7NM8jUoMzPLJAeUmZllkgPKzMwyyQFlZmaZ5IAyM7NMckCZmVkmOaDMzCyTHFBmZpZJ/wdKTkLOV9//VwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "content_stats = global_stats.loc[['Topics', 'Tutorials']]\n",
    "# plot barplot stats\n",
    "ax = content_stats.plot(kind='bar', legend=True, fontsize=12)\n",
    "ax.set_xticklabels(content_stats.index, rotation=45)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"../images/topic_stat_plot.png\", transparent=True, dpi=150)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Contributors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_contributors(ref):\n",
    "    '''\n",
    "    Extract the list of contributor at a given git commit\n",
    "\n",
    "    :param ref: name of the commit/branch/tag\n",
    "\n",
    "    :return: set with contributor ids\n",
    "    '''\n",
    "    file_content = repo.get_contents(\"CONTRIBUTORS.yaml\", ref=ref).content\n",
    "    decoded_file_content = base64.b64decode(file_content)\n",
    "    yaml_content = yaml.load(decoded_file_content, Loader=yaml.FullLoader)\n",
    "    contributors = set(yaml_content.keys())\n",
    "    return contributors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [],
   "source": [
    "contributors = {}\n",
    "contr_stats = pd.DataFrame(columns=['Number of contributors'], index=[])\n",
    "for d, git_com in gitcommit_dates.items():\n",
    "    if d == '22.06.17':\n",
    "        continue\n",
    "    contributors[d] = get_contributors(git_com)\n",
    "    contr_stats.loc[d] = {'Number of contributors': len(contributors[d])}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Number of contributors</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>22.06.18</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>22.06.19</td>\n",
       "      <td>122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>today</td>\n",
       "      <td>141</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Number of contributors\n",
       "22.06.18                     77\n",
       "22.06.19                    122\n",
       "today                       141"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contr_stats"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Contributions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 293,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "pr_stats = pd.DataFrame(\n",
    "    0,\n",
    "    columns=['PR', 'Reviews and comments (all PR)', 'Reviews and comments (mean per PR)', 'Reviewers (all PR)', 'Reviewers (mean per PR)', 'Duration (all PR)', 'Duration (mean per PR)'],\n",
    "    index=['PR to add tutorial', 'PR to update tutorial', 'Other PR'])\n",
    "pr_stats.loc[:, 'Duration (all PR)'] = datetime.timedelta(0)\n",
    "pr_stats.loc[:, 'Duration (mean per PR)'] = datetime.timedelta(0)\n",
    "prs = {} # keys: PR number as keys, values: PR object\n",
    "tuto_prs = {} # Keys: tutorials, values: set of PRs on the tutorial\n",
    "#n = 1\n",
    "for pr in repo.get_pulls(state=\"all\"):\n",
    "    # keep only merged PR\n",
    "    if not pr.merged:\n",
    "        continue\n",
    "    # try to identify PR for new tutorials\n",
    "    template_pr = False\n",
    "    status = 'Other PR'\n",
    "    for l in pr.labels:\n",
    "        if l.name == 'new tutorial':\n",
    "            status = 'PR to add tutorial'\n",
    "    prs[pr.number] = pr\n",
    "    # to test\n",
    "    #if n > 100:\n",
    "    #    break\n",
    "    #n += 1\n",
    "    # extract stats\n",
    "    review_nb = 0\n",
    "    comment_nb = 0\n",
    "    reviewers = set()\n",
    "    for r in pr.get_reviews():\n",
    "        reviewers.add(r.user.login)\n",
    "        review_nb += 1\n",
    "    for r in pr.get_issue_comments():\n",
    "        reviewers.add(r.user.login)\n",
    "        comment_nb += 1\n",
    "    # extract changed files\n",
    "    update = False\n",
    "    for f in pr.get_files():\n",
    "        if 'tutorials' in f.filename:\n",
    "            # update PR status\n",
    "            if status == 'Other PR':\n",
    "                status = 'PR to update tutorial'\n",
    "            if status != 'PR to add tutorial' and f.status == 'added': \n",
    "                if '.md' in f.filename or 'slides.html' in f.filename:\n",
    "                    status = 'PR to add tutorial'\n",
    "            # add tutorial to list of changed tutorials\n",
    "            tuto_n_m = re.search(r'([a-zA-Z0-9\\-_]+\\/tutorials\\/[a-zA-Z0-9\\-_]+)', f.filename)\n",
    "            if tuto_n_m:\n",
    "                tuto_n = tuto_n_m.group(0)\n",
    "                tuto_prs.setdefault(tuto_n, set())\n",
    "                tuto_prs[tuto_n].add(pr.number)\n",
    "    # add stats\n",
    "    pr_stats.loc[status, 'PR'] += 1\n",
    "    pr_stats.loc[status, 'Reviews and comments (all PR)'] += review_nb + comment_nb\n",
    "    pr_stats.loc[status, 'Reviewers (all PR)'] += len(reviewers)\n",
    "    pr_stats.loc[status, 'Duration (all PR)'] += (pr.merged_at-pr.created_at)\n",
    "# compute mean values\n",
    "pr_stats.loc[:, 'Reviews and comments (mean per PR)'] = pr_stats.loc[:,'Reviews and comments (all PR)']/pr_stats.loc[:, 'PR']\n",
    "pr_stats.loc[:, 'Reviewers (mean per PR)'] = pr_stats.loc[:,'Reviewers (all PR)']/pr_stats.loc[:, 'PR']\n",
    "pr_stats.loc[:, 'Duration (mean per PR)'] = pr_stats.loc[:,'Duration (all PR)']/pr_stats.loc[:, 'PR']\n",
    "# make copy\n",
    "all_tuto_prs = copy.deepcopy(tuto_prs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pull Request statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PR</th>\n",
       "      <th>Reviews and comments (all PR)</th>\n",
       "      <th>Reviews and comments (mean per PR)</th>\n",
       "      <th>Reviewers (all PR)</th>\n",
       "      <th>Reviewers (mean per PR)</th>\n",
       "      <th>Duration (all PR)</th>\n",
       "      <th>Duration (mean per PR)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>PR to add tutorial</td>\n",
       "      <td>133</td>\n",
       "      <td>1178</td>\n",
       "      <td>8.857143</td>\n",
       "      <td>344</td>\n",
       "      <td>2.586466</td>\n",
       "      <td>3247 days 23:58:29</td>\n",
       "      <td>24 days 10:06:18.263157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>PR to update tutorial</td>\n",
       "      <td>668</td>\n",
       "      <td>2244</td>\n",
       "      <td>3.359281</td>\n",
       "      <td>1147</td>\n",
       "      <td>1.717066</td>\n",
       "      <td>6639 days 11:14:31</td>\n",
       "      <td>9 days 22:32:37.591317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Other PR</td>\n",
       "      <td>377</td>\n",
       "      <td>923</td>\n",
       "      <td>2.448276</td>\n",
       "      <td>560</td>\n",
       "      <td>1.485411</td>\n",
       "      <td>1918 days 19:06:08</td>\n",
       "      <td>5 days 02:09:05.273209</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        PR  Reviews and comments (all PR)  \\\n",
       "PR to add tutorial     133                           1178   \n",
       "PR to update tutorial  668                           2244   \n",
       "Other PR               377                            923   \n",
       "\n",
       "                       Reviews and comments (mean per PR)  Reviewers (all PR)  \\\n",
       "PR to add tutorial                               8.857143                 344   \n",
       "PR to update tutorial                            3.359281                1147   \n",
       "Other PR                                         2.448276                 560   \n",
       "\n",
       "                       Reviewers (mean per PR)  Duration (all PR)  \\\n",
       "PR to add tutorial                    2.586466 3247 days 23:58:29   \n",
       "PR to update tutorial                 1.717066 6639 days 11:14:31   \n",
       "Other PR                              1.485411 1918 days 19:06:08   \n",
       "\n",
       "                       Duration (mean per PR)  \n",
       "PR to add tutorial    24 days 10:06:18.263157  \n",
       "PR to update tutorial  9 days 22:32:37.591317  \n",
       "Other PR               5 days 02:09:05.273209  "
      ]
     },
     "execution_count": 294,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pr_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 296,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['dev/tutorials/conda',\n",
       " 'epigenetics/tutorials/atac-seq',\n",
       " 'transcriptomics/tutorials/rna-seq-reads-to-counts',\n",
       " 'galaxy-data-manipulation/tutorials/get-data',\n",
       " 'transcriptomics/tutorials/srna',\n",
       " 'transcriptomics/tutorials/ref-based',\n",
       " 'sequence-analysis/tutorials/mapping',\n",
       " 'computational-chemistry/tutorials/cheminformatics',\n",
       " 'computational-chemistry/tutorials/md-simulation-gromacs',\n",
       " 'admin/tutorials/ansible-galaxy',\n",
       " 'metagenomics/tutorials/general-tutorial',\n",
       " 'transcriptomics/tutorials/rna-seq-counts-to-viz-in-r',\n",
       " 'galaxy-data-manipulation/tutorials/download-delete-data',\n",
       " 'introduction/tutorials/galaxy-intro-101',\n",
       " 'transcriptomics/tutorials/scrna-preprocessing-tenx',\n",
       " 'dev/tutorials/data-source-integration',\n",
       " 'variant-analysis/tutorials/exome-seq',\n",
       " 'variant-analysis/tutorials/somatic-variants',\n",
       " 'metagenomics/tutorials/mothur-miseq-sop',\n",
       " 'epigenetics/tutorials/estrogen-receptor-binding-site-identification',\n",
       " 'epigenetics/tutorials/formation_of_super-structures_on_xi',\n",
       " 'epigenetics/tutorials/tal1-binding-site-identification',\n",
       " 'transcriptomics/tutorials/clipseq',\n",
       " 'ecology/tutorials/de-novo-rad-seq',\n",
       " 'ecology/tutorials/genetic-map-rad-seq',\n",
       " 'ecology/tutorials/ref-based-rad-seq',\n",
       " 'transcriptomics/tutorials/scrna-preprocessing',\n",
       " 'proteomics/tutorials/metaproteomics',\n",
       " 'variant-analysis/tutorials/dunovo',\n",
       " 'epigenetics/tutorials/ewas-suite',\n",
       " 'transcriptomics/tutorials/scrna-scater-qc',\n",
       " 'metabolomics/tutorials/lcms',\n",
       " 'ecology/tutorials/regionalGAM',\n",
       " 'contributing/tutorials/create-new-topic',\n",
       " 'transcriptomics/tutorials/scrna-plates-batches-barcodes',\n",
       " 'transcriptomics/tutorials/scrna-raceid',\n",
       " 'transcriptomics/tutorials/scrna-umis',\n",
       " 'instructors/tutorials/organize-workshop',\n",
       " 'instructors/tutorials/running-workshop',\n",
       " 'admin/tutorials/jenkins',\n",
       " 'imaging/tutorials/hela-screen-analysis',\n",
       " 'imaging/tutorials/imaging-introduction',\n",
       " 'admin/tutorials/monitoring',\n",
       " 'sequence-analysis/tutorials/quality-control',\n",
       " 'dev/tutorials/bioblend-api',\n",
       " 'proteomics/tutorials/biomarker_selection',\n",
       " 'contributing/tutorials/create-new-tutorial',\n",
       " 'admin/tutorials/ansible',\n",
       " 'admin/tutorials/connect-to-compute-cluster',\n",
       " 'admin/tutorials/cvmfs',\n",
       " 'admin/tutorials/external-auth',\n",
       " 'admin/tutorials/gxadmin',\n",
       " 'admin/tutorials/heterogeneous-compute',\n",
       " 'admin/tutorials/object-store',\n",
       " 'admin/tutorials/reports',\n",
       " 'admin/tutorials/terraform',\n",
       " 'admin/tutorials/tool-management',\n",
       " 'contributing/tutorials/create-new-tutorial-content',\n",
       " 'galaxy-ui/tutorials/rstudio',\n",
       " 'introduction/tutorials/r-basics',\n",
       " 'proteomics/tutorials/proteogenomics-novel-peptide-analysis',\n",
       " 'proteomics/tutorials/database-handling',\n",
       " 'proteomics/tutorials/protein-quant-sil',\n",
       " 'proteomics/tutorials/protein-id-oms',\n",
       " 'proteomics/tutorials/mass-spectrometry-imaging-loading-exploring-data',\n",
       " 'metabolomics/tutorials/msi-finding-nglycans',\n",
       " 'metabolomics/tutorials/msi-analyte-distribution',\n",
       " 'assembly/tutorials/general-introduction',\n",
       " 'admin/tutorials/k8s-deploying-galaxy',\n",
       " 'admin/tutorials/k8s-managing-galaxy',\n",
       " 'transcriptomics/tutorials/rna-seq-viz-with-volcanoplot',\n",
       " 'assembly/tutorials/debruijn-graph-assembly',\n",
       " 'assembly/tutorials/ecoli_comparison',\n",
       " 'assembly/tutorials/unicycler-assembly',\n",
       " 'epigenetics/tutorials/hicexplorer',\n",
       " 'epigenetics/tutorials/methylation-seq',\n",
       " 'genome-annotation/tutorials/annotation-with-maker',\n",
       " 'genome-annotation/tutorials/annotation-with-prokka',\n",
       " 'introduction/tutorials/galaxy-intro-short',\n",
       " 'introduction/tutorials/igv-introduction',\n",
       " 'introduction/tutorials/options-for-using-galaxy',\n",
       " 'metagenomics/tutorials/plasmid-metagenomics-nanopore',\n",
       " 'proteomics/tutorials/protein-id-sg-ps',\n",
       " 'proteomics/tutorials/proteogenomics-dbcreation',\n",
       " 'proteomics/tutorials/proteogenomics-dbsearch',\n",
       " 'proteomics/tutorials/secretome-prediction',\n",
       " 'sequence-analysis/tutorials/de-novo-rad-seq',\n",
       " 'sequence-analysis/tutorials/genetic-map-rad-seq',\n",
       " 'sequence-analysis/tutorials/ref-based-rad-seq',\n",
       " 'statistics/tutorials/iwtomics',\n",
       " 'transcriptomics/tutorials/de-novo',\n",
       " 'transcriptomics/tutorials/goenrichment',\n",
       " 'transcriptomics/tutorials/network-analysis-with-heinz',\n",
       " 'transcriptomics/tutorials/rna-seq-viz-with-cummerbund',\n",
       " 'variant-analysis/tutorials/dip',\n",
       " 'variant-analysis/tutorials/mapping-by-sequencing',\n",
       " 'variant-analysis/tutorials/microbial-variants',\n",
       " 'transcriptomics/tutorials/scrna_preprocessing',\n",
       " 'dev/tutorials/tool-integration',\n",
       " 'admin/tutorials/uwsgi',\n",
       " 'instructors/tutorials/setup-galaxy-for-training',\n",
       " 'galaxy-data-manipulation/tutorials/upload-rules',\n",
       " 'dev/tutorials/visualization-charts',\n",
       " 'instructors/tutorials/setup-tiaas-for-training',\n",
       " 'proteomics/tutorials/proteome_annotation',\n",
       " 'statistics/tutorials/classification_regression',\n",
       " 'statistics/tutorials/age-prediction-with-ml',\n",
       " 'metabolomics/tutorials/lcms-data-import',\n",
       " 'computational-chemistry/tutorials/analysis-md-simulations',\n",
       " 'galaxy-data-manipulation/tutorials/group-tags',\n",
       " 'introduction/tutorials/galaxy-intro-peaks2genes',\n",
       " 'statistics/tutorials/machinelearning',\n",
       " 'transcriptomics/tutorials/rna-seq-counts-to-genes',\n",
       " 'transcriptomics/tutorials/rna-seq-viz-with-heatmap2',\n",
       " 'transcriptomics/tutorials/small_ncrna_clustering',\n",
       " 'instructors/tutorials/workshop-intro',\n",
       " 'admin/tutorials/production',\n",
       " 'galaxy-ui/tutorials/workflow-parameters',\n",
       " 'metagenomics/tutorials/mothur-miseq-sop-short',\n",
       " 'computational-chemistry/tutorials/md-simulation-namd',\n",
       " 'computational-chemistry/tutorials/setting-up-molecular-systems',\n",
       " 'galaxy-data-manipulation/tutorials/collections',\n",
       " 'dev/tutorials/visualization-generic',\n",
       " 'galaxy-data-manipulation/tutorials/name-tags',\n",
       " 'dev/tutorials/toolshed',\n",
       " 'transcriptomics/tutorials/rna-seq-genes-to-pathways',\n",
       " 'introduction/tutorials/galaxy-intro-strands',\n",
       " 'contributing/tutorials/create-new-tutorial-technical',\n",
       " 'admin/tutorials/upstream-auth',\n",
       " 'contributing/tutorials/github-command-line-contribution',\n",
       " 'contributing/tutorials/github-interface-contribution',\n",
       " 'contributing/tutorials/running-jekyll',\n",
       " 'galaxy-ui/tutorials/history-to-workflow',\n",
       " 'genome-annotation/tutorials/genome-annotation',\n",
       " 'galaxy-data-manipulation/tutorials/processing-many-samples-at-once',\n",
       " 'genome-annotation/tutorials/tnseq',\n",
       " 'proteomics/tutorials/labelfree-vs-labelled',\n",
       " 'proteomics/tutorials/ntails',\n",
       " 'admin/tutorials/cloudbursting',\n",
       " 'admin/tutorials/reference-genomes',\n",
       " 'admin/tutorials/troubleshooting',\n",
       " 'contributing/tutorials/create-new-tutorial-slides',\n",
       " 'introduction/tutorials/galaxy-intro-ngs-data-managment',\n",
       " 'variant-analysis/tutorials/non-dip',\n",
       " 'dev/tutorials/containers',\n",
       " 'dev/tutorials/interactive-environments',\n",
       " 'admin/tutorials/galaxy-docker',\n",
       " 'dev/tutorials/interactive-tours',\n",
       " 'dev/tutorials/webhooks',\n",
       " 'galaxy-ui/tutorials/galaxy-intro-jupyter',\n",
       " 'dev/tutorials/architecture',\n",
       " 'admin/tutorials/advanced-galaxy-customisation',\n",
       " 'admin/tutorials/deployment-platforms-options',\n",
       " 'admin/tutorials/general-monitoring',\n",
       " 'admin/tutorials/job-conf',\n",
       " 'admin/tutorials/maintenance',\n",
       " 'admin/tutorials/tool-install',\n",
       " 'admin/tutorials/webservers',\n",
       " 'admin/tutorials/upgrading',\n",
       " 'admin/tutorials/stuff',\n",
       " 'admin/tutorials/empathy',\n",
       " 'admin/tutorials/database',\n",
       " 'admin/tutorials/dev-to-production',\n",
       " 'admin/tutorials/systemd-supervisor',\n",
       " 'admin/tutorials/toolshed',\n",
       " 'admin/tutorials/users-groups-quotas',\n",
       " 'transcriptomics/tutorials/limma-voom',\n",
       " 'transcriptomics/tutorials/limma-voom_fastqs_to_counts',\n",
       " 'transcriptomics/tutorials/rb-rnaseq',\n",
       " 'ecology/tutorials/species-distribution-modeling',\n",
       " 'contributing/tutorials/good-practices',\n",
       " 'contributing/tutorials/generating-pdf',\n",
       " 'contributing/tutorials/create-new-tutorial-tours',\n",
       " 'galaxy-ui/tutorials/history',\n",
       " 'admin/tutorials/database-schema',\n",
       " 'dev/tutorials/conda_sys',\n",
       " 'variant-analysis/tutorials/diploid-variant-calling',\n",
       " 'introduction/tutorials/galaxy-intro-vis',\n",
       " 'templates/tutorials/tutorial1',\n",
       " 'contributing/tutorials/create-new-tutorial-metadata',\n",
       " 'introduction/tutorials/galaxy-intro-collections',\n",
       " 'introduction/tutorials/galaxy-intro-get-data',\n",
       " 'introduction/tutorials/processing-many-samples-at-once',\n",
       " 'introduction/tutorials/galaxy-intro-rules',\n",
       " 'training/tutorials/create-new-tutorial-content',\n",
       " 'training/tutorials/create-new-tutorial-docker',\n",
       " 'training/tutorials/create-new-tutorial-jekyll',\n",
       " 'training/tutorials/create-new-tutorial-metadata',\n",
       " 'training/tutorials/create-new-tutorial-tours',\n",
       " 'sequence-analysis/tutorials/genome-annotation',\n",
       " 'sequence-analysis/tutorials/annotation-with-prokka',\n",
       " 'dev/tutorials/github-contribution',\n",
       " 'introduction/tutorials/galaxy-intro-history',\n",
       " 'usegalaxy/tutorials/collections',\n",
       " 'introduction/tutorials/galaxy-intro-history-to-workflow',\n",
       " 'usegalaxy/tutorials/dip',\n",
       " 'training/tutorials/good-practices',\n",
       " 'usegalaxy/tutorials/rb-rnaseq',\n",
       " 'usegalaxy/tutorials/ngs',\n",
       " 'usegalaxy/tutorials/dunovo',\n",
       " 'usegalaxy/tutorials/history',\n",
       " 'usegalaxy/tutorials/non-dip',\n",
       " 'sequence-analysis/tutorials/srna',\n",
       " 'Admin-Corner/tutorials/dev_to_production',\n",
       " 'Admin-Corner/tutorials/mysql_to_postgresql',\n",
       " 'Admin-Corner/tutorials/setting_up_galaxy_instance',\n",
       " 'Introduction/tutorials/IGV_Introduction',\n",
       " 'Introduction/tutorials/Processing-many-samples-at-once',\n",
       " 'Introduction/tutorials/galaxy-intro-101',\n",
       " 'Introduction/tutorials/galaxy-intro-collections',\n",
       " 'Introduction/tutorials/galaxy-intro-get-data',\n",
       " 'Introduction/tutorials/galaxy-intro-peaks2genes',\n",
       " 'Introduction/tutorials/galaxy-intro-vis',\n",
       " 'Metagenomics/tutorials/mothur-miseq-sop',\n",
       " 'MethylC-Seq/tutorials/methylation-seq',\n",
       " 'NGS-QC/tutorials/dive_into_qc',\n",
       " 'NGS-mapping/tutorials/dive_into_mapping',\n",
       " 'Proteomics/tutorials/database-handling',\n",
       " 'Proteomics/tutorials/labelfree_vs_labelled',\n",
       " 'Proteomics/tutorials/proteinID_SG_PS',\n",
       " 'Proteomics/tutorials/proteinQuant_SIL',\n",
       " 'RAD-Seq/tutorials/de_novo_rad',\n",
       " 'RAD-Seq/tutorials/genetic_map_rad',\n",
       " 'RAD-Seq/tutorials/ref_based_rad',\n",
       " 'RNA-Seq/tutorials/de_novo',\n",
       " 'RNA-Seq/tutorials/ref_based',\n",
       " 'RNA-Seq/tutorials/srna-seq',\n",
       " 'Usegalaxy/tutorials/collections',\n",
       " 'Usegalaxy/tutorials/dip',\n",
       " 'Usegalaxy/tutorials/dunovo',\n",
       " 'Usegalaxy/tutorials/history',\n",
       " 'Usegalaxy/tutorials/ngs',\n",
       " 'Usegalaxy/tutorials/non_dip',\n",
       " 'Usegalaxy/tutorials/rb_rnaseq',\n",
       " 'templates/tutorials/tutorial',\n",
       " 'Assembly/tutorials/index',\n",
       " 'Assembly/tutorials/tutorial',\n",
       " 'Introduction/tutorials/Galaxy101-1',\n",
       " 'Introduction/tutorials/Galaxy101-2',\n",
       " 'Introduction/tutorials/Introduction_to_Galaxy_Uni',\n",
       " 'Introduction/tutorials/introduction_my_first_analysis',\n",
       " 'Proteomics/tutorials/WIP_proteinQuant_SIL',\n",
       " 'MethylC-Seq/tutorials/Methylation-Seq',\n",
       " 'Usegalaxy/tutorials/cp_img',\n",
       " 'Introduction/tutorials/Introduction',\n",
       " 'Introduction/tutorials/galaxy-introduction',\n",
       " 'Dev-Corner/tutorials/files',\n",
       " 'MethylC-Seq/tutorials/methylc_seq',\n",
       " 'RNA-Seq/tutorials/RNA-Seq-hands-on',\n",
       " 'RNA-Seq/tutorials/Reference-based-RNA-seq-detailed',\n",
       " 'Introduction/tutorials/images',\n",
       " 'Introduction/tutorials/igv',\n",
       " 'Introduction/tutorials/processing_many_samples',\n",
       " 'Introduction/tutorials/workflows']"
      ]
     },
     "execution_count": 296,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(tuto_prs.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 301,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "164"
      ]
     },
     "execution_count": 301,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tuto_prs = copy.deepcopy(all_tuto_prs)\n",
    "# merge some tutorials that have been moved / renamed\n",
    "to_merge = [\n",
    "    ['admin/tutorials/heterogeneous', 'admin/tutorials/heterogeneous-compute'],\n",
    "    ['admin/tutorials/monitoring-maintenance', 'admin/tutorials/monitoring'],\n",
    "    ['Admin-Corner/tutorials/Database_Schema', 'admin/tutorials/database-schema'],\n",
    "    ['Admin-Corner/tutorials/galaxy_docker', 'admin/tutorials/galaxy-docker'],\n",
    "    \n",
    "    ['Assembly/tutorials/debruijn-graph-assembly', 'assembly/tutorials/debruijn-graph-assembly'],\n",
    "    ['Assembly/tutorials/general-introduction', 'assembly/tutorials/general-introduction'],\n",
    "\n",
    "    ['chip-seq/tutorials/chip-seq', 'epigenetics/tutorials/estrogen-receptor-binding-site-identification'],\n",
    "    ['chip-seq/tutorials/ChIPseq', 'epigenetics/tutorials/estrogen-receptor-binding-site-identification'],\n",
    "    ['chip-seq/tutorials/estrogen-receptor-binding-site-identification', 'epigenetics/tutorials/estrogen-receptor-binding-site-identification'],\n",
    "    ['chip-seq/tutorials/formation_of_super-structures_on_xi', 'epigenetics/tutorials/formation_of_super-structures_on_xi'],\n",
    "    ['chip-seq/tutorials/tal1-binding-site-identification', 'epigenetics/tutorials/tal1-binding-site-identification'],\n",
    "    ['ChIP-Seq/tutorials/ChIPseq', 'epigenetics/tutorials/estrogen-receptor-binding-site-identification'],\n",
    "    ['ChIP-Seq/tutorials/estrogen_receptor_binding_site_identification', 'epigenetics/tutorials/estrogen-receptor-binding-site-identification'],\n",
    "    ['ChIP-Seq/tutorials/TAL1_binding_site_identification', 'epigenetics/tutorials/tal1-binding-site-identification'],\n",
    "    \n",
    "    ['contributing/tutorials/setup-galaxy-for-training', 'instructors/tutorials/setup-galaxy-for-training'],\n",
    "\n",
    "    ['Dev-Corner/tutorials/data_source_integration', 'dev/tutorials/data-source-integration'],\n",
    "    ['Dev-Corner/tutorials/tool_integration', 'dev/tutorials/tool-integration'],\n",
    "    ['Dev-Corner/tutorials/visualizations', 'dev/tutorials/visualization-generic'],\n",
    "    ['dev/tutorials/visualization', 'dev/tutorials/visualization-generic'],\n",
    "    ['dev/tutorials/interactive-environment', 'dev/tutorials/interactive-environments'],\n",
    "    \n",
    "    ['Exome-Seq/tutorials/Diploid-variant-calling', 'variant-analysis/tutorials/dip'],\n",
    "    ['Exome-Seq/tutorials/Exome-Seq', 'variant-analysis/tutorials/exome-seq'],\n",
    "    \n",
    "    ['Genome-Annotation/tutorials/general-introduction', 'genome-annotation/tutorials/genome-annotation'],\n",
    "    \n",
    "    ['introduction/tutorials/galaxy-intro-collections', 'galaxy-data-manipulation/tutorials/collections'],\n",
    "    ['introduction/tutorials/galaxy-intro-get-data', 'galaxy-data-manipulation/tutorials/get-data'],\n",
    "    ['introduction/tutorials/galaxy-intro-rules', 'galaxy-data-manipulation/tutorials/upload-rules'],\n",
    "    ['introduction/tutorials/galaxy-intro-history', 'galaxy-ui/tutorials/history'],\n",
    "    ['introduction/tutorials/galaxy-intro-history-to-workflow', 'galaxy-ui/tutorials/history-to-workflow'],\n",
    "    ['introduction/tutorials/processing-many-samples-at-once', 'galaxy-data-manipulation/tutorials/processing-many-samples-at-once'],\n",
    "    ['Introduction/tutorials/galaxy-intro-101', 'introduction/tutorials/galaxy-intro-101'],\n",
    "    ['Introduction/tutorials/galaxy-intro-collections', 'galaxy-data-manipulation/tutorials/collections'],\n",
    "    ['Introduction/tutorials/galaxy-intro-get-data', 'galaxy-data-manipulation/tutorials/get-data'],\n",
    "    ['Introduction/tutorials/galaxy-intro-peaks2genes', 'introduction/tutorials/galaxy-intro-peaks2genes'],\n",
    "    ['Introduction/tutorials/Galaxy101-1', 'introduction/tutorials/galaxy-intro-101'],\n",
    "    ['Introduction/tutorials/Galaxy101-2', 'introduction/tutorials/galaxy-intro-101'],\n",
    "    ['Introduction/tutorials/igv', 'introduction/tutorials/igv-introduction'],\n",
    "    ['Introduction/tutorials/IGV_Introduction', 'introduction/tutorials/igv-introduction'],\n",
    "    ['Introduction/tutorials/processing_many_samples', 'galaxy-data-manipulation/tutorials/processing-many-samples-at-once'],\n",
    "    ['Introduction/tutorials/Processing-many-samples-at-once', 'galaxy-data-manipulation/tutorials/processing-many-samples-at-once'],\n",
    "    ['Introduction/tutorials/workflows', 'galaxy-ui/tutorials/workflow-parameters'],\n",
    "    \n",
    "    ['Metagenomics/tutorials/mothur-miseq-sop', 'metagenomics/tutorials/mothur-miseq-sop'],\n",
    "    \n",
    "    ['MethylC-Seq/tutorials/methylation-seq', 'epigenetics/tutorials/methylation-seq'],\n",
    "    ['MethylC-Seq/tutorials/methylc_seq', 'epigenetics/tutorials/methylation-seq'],\n",
    "    ['MethylC-Seq/tutorials/Methylation-Seq', 'epigenetics/tutorials/methylation-seq'],\n",
    "    \n",
    "    ['NGS-QC/tutorials/dive_into_qc', 'sequence-analysis/tutorials/quality-control'],\n",
    "    \n",
    "    ['NGS-mapping/tutorials/dive_into_mapping', 'sequence-analysis/tutorials/mapping'],\n",
    "    \n",
    "    ['Proteomics/tutorials/WIP_proteinQuant_SIL', 'proteomics/tutorials/protein-quant-sil'],\n",
    "    ['Proteomics/tutorials/database-handling', 'proteomics/tutorials/database-handling'],\n",
    "    ['Proteomics/tutorials/labelfree_vs_labelled', 'proteomics/tutorials/labelfree-vs-labelled'],\n",
    "    ['Proteomics/tutorials/proteinID_SG_PS', 'proteomics/tutorials/protein-id-sg-ps'],\n",
    "    ['Proteomics/tutorials/proteinQuant_SIL', 'proteomics/tutorials/protein-quant-sil'],\n",
    "    \n",
    "    ['RAD-Seq/tutorials/de_novo_rad', 'ecology/tutorials/de-novo-rad-seq'],\n",
    "    ['RAD-Seq/tutorials/genetic_map_rad', 'ecology/tutorials/genetic-map-rad-seq'],\n",
    "    ['RAD-Seq/tutorials/ref_based_rad', 'ecology/tutorials/ref-based-rad-seq'],\n",
    "    \n",
    "    ['RNA-Seq/tutorials/de_novo', 'transcriptomics/tutorials/de-novo'],\n",
    "    ['RNA-Seq/tutorials/ref_based', 'transcriptomics/tutorials/ref-based'],\n",
    "    ['RNA-Seq/tutorials/srna-seq', 'transcriptomics/tutorials/srna'],\n",
    "    ['RNA-Seq/tutorials/Reference-based-RNA-seq-detailed', 'transcriptomics/tutorials/ref-based'],\n",
    "    ['RNA-Seq/tutorials/RNA-Seq-hands-on', 'transcriptomics/tutorials/ref-based'],\n",
    "    \n",
    "    ['sequence-analysis/tutorials/ref-based-rad-seq', 'ecology/tutorials/ref-based-rad-seq'],\n",
    "    ['sequence-analysis/tutorials/de-novo-rad-seq', 'ecology/tutorials/de-novo-rad-seq'],\n",
    "    ['sequence-analysis/tutorials/genetic-map-rad-seq', 'ecology/tutorials/genetic-map-rad-seq'],\n",
    "    ['sequence-analysis/tutorials/genome-annotation', 'genome-annotation/tutorials/genome-annotation'],\n",
    "    ['sequence-analysis/tutorials/annotation-with-prokka', 'genome-annotation/tutorials/annotation-with-prokka' ],\n",
    "    ['sequence-analysis/tutorials/srna', 'transcriptomics/tutorials/srna'],\n",
    "    \n",
    "    ['training/tutorials/create-new-tutorial-content', 'contributing/tutorials/create-new-tutorial-content'],\n",
    "    ['training/tutorials/create-new-tutorial-tours', 'contributing/tutorials/create-new-tutorial-tours'],\n",
    "    ['training/tutorials/create-new-tutorial-jekyll', 'contributing/tutorials/running-jekyll'],\n",
    "    \n",
    "    ['transcriptomics/tutorials/scrna_preprocessing', 'transcriptomics/tutorials/scrna-preprocessing'],\n",
    "    ['transcriptomics/tutorials/limma-voom', 'transcriptomics/tutorials/rna-seq-counts-to-genes',],\n",
    "    ['transcriptomics/tutorials/limma-voom_fastqs_to_counts', 'transcriptomics/tutorials/rna-seq-reads-to-counts'],\n",
    "    \n",
    "    ['Usegalaxy/tutorials/collections', 'galaxy-data-manipulation/tutorials/collections'],\n",
    "    ['Usegalaxy/tutorials/dip', 'variant-analysis/tutorials/dip'],\n",
    "    ['Usegalaxy/tutorials/dunovo', 'variant-analysis/tutorials/dunovo'],\n",
    "    ['Usegalaxy/tutorials/history', 'galaxy-ui/tutorials/history'],\n",
    "    ['Usegalaxy/tutorials/ngs', 'introduction/tutorials/galaxy-intro-ngs-data-managment'],\n",
    "    ['Usegalaxy/tutorials/non_dip', 'variant-analysis/tutorials/non-dip'],\n",
    "    ['Usegalaxy/tutorials/rb_rnaseq', 'transcriptomics/tutorials/rb-rnaseq'],\n",
    "    ['usegalaxy/tutorials/collections', 'galaxy-data-manipulation/tutorials/collections'],\n",
    "    ['usegalaxy/tutorials/dip', 'variant-analysis/tutorials/dip'],\n",
    "    ['usegalaxy/tutorials/rb-rnaseq', 'transcriptomics/tutorials/rb-rnaseq'],\n",
    "    ['usegalaxy/tutorials/ngs', 'introduction/tutorials/galaxy-intro-ngs-data-managment'],\n",
    "    ['usegalaxy/tutorials/dunovo', 'variant-analysis/tutorials/dunovo'],\n",
    "    ['usegalaxy/tutorials/history', 'galaxy-ui/tutorials/history'],\n",
    "    ['usegalaxy/tutorials/non-dip', 'variant-analysis/tutorials/non-dip'],\n",
    "\n",
    "    ['variant-analysis/tutorials/diploid-variant-calling', 'variant-analysis/tutorials/dip'],\n",
    "]\n",
    "for t in to_merge:\n",
    "    tuto_prs[t[1]].update(tuto_prs[t[0]])\n",
    "    del tuto_prs[t[0]]\n",
    "    \n",
    "    \n",
    "to_remove = [\n",
    "    'Introduction/tutorials/images',\n",
    "    'Usegalaxy/tutorials/cp_img',\n",
    "    'templates/tutorials/tutorial',\n",
    "    'templates/tutorials/tutorial1',\n",
    "    \n",
    "    'Assembly/tutorials/index',\n",
    "    'Assembly/tutorials/tutorial',\n",
    "    'Dev-Corner/tutorials/files'\n",
    "    \n",
    "    # removed tutorials\n",
    "    'admin/tutorials/deployment-platforms-options',\n",
    "    'admin/tutorials/dev-to-production',\n",
    "    'admin/tutorials/job-conf',\n",
    "    'admin/tutorials/webservers',\n",
    "    'admin/tutorials/upstream-auth',\n",
    "    'Admin-Corner/tutorials/dev_to_production',\n",
    "    'Admin-Corner/tutorials/mysql_to_postgresql',\n",
    "    'Admin-Corner/tutorials/setting_up_galaxy_instance',\n",
    "    'contributing/tutorials/create-new-tutorial-metadata',\n",
    "    'contributing/tutorials/good-practices',\n",
    "    'dev/tutorials/github-contribution',\n",
    "    'introduction/tutorials/galaxy-intro-vis',\n",
    "    'Introduction/tutorials/galaxy-introduction',\n",
    "    'Introduction/tutorials/galaxy-intro-vis',\n",
    "    'Introduction/tutorials/introduction_my_first_analysis',\n",
    "    'Introduction/tutorials/Introduction',\n",
    "    'Introduction/tutorials/Introduction_to_Galaxy_Uni',\n",
    "    'metabolomics/tutorials/lcms-data-import',\n",
    "    'training/tutorials/create-new-tutorial-docker', \n",
    "    'training/tutorials/create-new-tutorial-metadata', \n",
    "    'training/tutorials/good-practices', \n",
    "]\n",
    "for t in to_remove:\n",
    "    del tuto_prs[t]\n",
    "    \n",
    "list(tuto_prs.keys())\n",
    "len(tuto_prs.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 303,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['dev/tutorials/conda',\n",
       " 'epigenetics/tutorials/atac-seq',\n",
       " 'transcriptomics/tutorials/rna-seq-reads-to-counts',\n",
       " 'galaxy-data-manipulation/tutorials/get-data',\n",
       " 'transcriptomics/tutorials/srna',\n",
       " 'transcriptomics/tutorials/ref-based',\n",
       " 'sequence-analysis/tutorials/mapping',\n",
       " 'computational-chemistry/tutorials/cheminformatics',\n",
       " 'computational-chemistry/tutorials/md-simulation-gromacs',\n",
       " 'admin/tutorials/ansible-galaxy',\n",
       " 'metagenomics/tutorials/general-tutorial',\n",
       " 'transcriptomics/tutorials/rna-seq-counts-to-viz-in-r',\n",
       " 'galaxy-data-manipulation/tutorials/download-delete-data',\n",
       " 'introduction/tutorials/galaxy-intro-101',\n",
       " 'transcriptomics/tutorials/scrna-preprocessing-tenx',\n",
       " 'dev/tutorials/data-source-integration',\n",
       " 'variant-analysis/tutorials/exome-seq',\n",
       " 'variant-analysis/tutorials/somatic-variants',\n",
       " 'metagenomics/tutorials/mothur-miseq-sop',\n",
       " 'epigenetics/tutorials/estrogen-receptor-binding-site-identification',\n",
       " 'epigenetics/tutorials/formation_of_super-structures_on_xi',\n",
       " 'epigenetics/tutorials/tal1-binding-site-identification',\n",
       " 'transcriptomics/tutorials/clipseq',\n",
       " 'ecology/tutorials/de-novo-rad-seq',\n",
       " 'ecology/tutorials/genetic-map-rad-seq',\n",
       " 'ecology/tutorials/ref-based-rad-seq',\n",
       " 'transcriptomics/tutorials/scrna-preprocessing',\n",
       " 'proteomics/tutorials/metaproteomics',\n",
       " 'variant-analysis/tutorials/dunovo',\n",
       " 'epigenetics/tutorials/ewas-suite',\n",
       " 'transcriptomics/tutorials/scrna-scater-qc',\n",
       " 'metabolomics/tutorials/lcms',\n",
       " 'ecology/tutorials/regionalGAM',\n",
       " 'contributing/tutorials/create-new-topic',\n",
       " 'transcriptomics/tutorials/scrna-plates-batches-barcodes',\n",
       " 'transcriptomics/tutorials/scrna-raceid',\n",
       " 'transcriptomics/tutorials/scrna-umis',\n",
       " 'instructors/tutorials/organize-workshop',\n",
       " 'instructors/tutorials/running-workshop',\n",
       " 'admin/tutorials/jenkins',\n",
       " 'imaging/tutorials/hela-screen-analysis',\n",
       " 'imaging/tutorials/imaging-introduction',\n",
       " 'admin/tutorials/monitoring',\n",
       " 'sequence-analysis/tutorials/quality-control',\n",
       " 'dev/tutorials/bioblend-api',\n",
       " 'proteomics/tutorials/biomarker_selection',\n",
       " 'contributing/tutorials/create-new-tutorial',\n",
       " 'admin/tutorials/ansible',\n",
       " 'admin/tutorials/connect-to-compute-cluster',\n",
       " 'admin/tutorials/cvmfs',\n",
       " 'admin/tutorials/external-auth',\n",
       " 'admin/tutorials/gxadmin',\n",
       " 'admin/tutorials/heterogeneous-compute',\n",
       " 'admin/tutorials/object-store',\n",
       " 'admin/tutorials/reports',\n",
       " 'admin/tutorials/terraform',\n",
       " 'admin/tutorials/tool-management',\n",
       " 'contributing/tutorials/create-new-tutorial-content',\n",
       " 'galaxy-ui/tutorials/rstudio',\n",
       " 'introduction/tutorials/r-basics',\n",
       " 'proteomics/tutorials/proteogenomics-novel-peptide-analysis',\n",
       " 'proteomics/tutorials/database-handling',\n",
       " 'proteomics/tutorials/protein-quant-sil',\n",
       " 'proteomics/tutorials/protein-id-oms',\n",
       " 'proteomics/tutorials/mass-spectrometry-imaging-loading-exploring-data',\n",
       " 'metabolomics/tutorials/msi-finding-nglycans',\n",
       " 'metabolomics/tutorials/msi-analyte-distribution',\n",
       " 'assembly/tutorials/general-introduction',\n",
       " 'admin/tutorials/k8s-deploying-galaxy',\n",
       " 'admin/tutorials/k8s-managing-galaxy',\n",
       " 'transcriptomics/tutorials/rna-seq-viz-with-volcanoplot',\n",
       " 'assembly/tutorials/debruijn-graph-assembly',\n",
       " 'assembly/tutorials/ecoli_comparison',\n",
       " 'assembly/tutorials/unicycler-assembly',\n",
       " 'epigenetics/tutorials/hicexplorer',\n",
       " 'epigenetics/tutorials/methylation-seq',\n",
       " 'genome-annotation/tutorials/annotation-with-maker',\n",
       " 'genome-annotation/tutorials/annotation-with-prokka',\n",
       " 'introduction/tutorials/galaxy-intro-short',\n",
       " 'introduction/tutorials/igv-introduction',\n",
       " 'introduction/tutorials/options-for-using-galaxy',\n",
       " 'metagenomics/tutorials/plasmid-metagenomics-nanopore',\n",
       " 'proteomics/tutorials/protein-id-sg-ps',\n",
       " 'proteomics/tutorials/proteogenomics-dbcreation',\n",
       " 'proteomics/tutorials/proteogenomics-dbsearch',\n",
       " 'proteomics/tutorials/secretome-prediction',\n",
       " 'statistics/tutorials/iwtomics',\n",
       " 'transcriptomics/tutorials/de-novo',\n",
       " 'transcriptomics/tutorials/goenrichment',\n",
       " 'transcriptomics/tutorials/network-analysis-with-heinz',\n",
       " 'transcriptomics/tutorials/rna-seq-viz-with-cummerbund',\n",
       " 'variant-analysis/tutorials/dip',\n",
       " 'variant-analysis/tutorials/mapping-by-sequencing',\n",
       " 'variant-analysis/tutorials/microbial-variants',\n",
       " 'dev/tutorials/tool-integration',\n",
       " 'admin/tutorials/uwsgi',\n",
       " 'instructors/tutorials/setup-galaxy-for-training',\n",
       " 'galaxy-data-manipulation/tutorials/upload-rules',\n",
       " 'dev/tutorials/visualization-charts',\n",
       " 'instructors/tutorials/setup-tiaas-for-training',\n",
       " 'proteomics/tutorials/proteome_annotation',\n",
       " 'statistics/tutorials/classification_regression',\n",
       " 'statistics/tutorials/age-prediction-with-ml',\n",
       " 'computational-chemistry/tutorials/analysis-md-simulations',\n",
       " 'galaxy-data-manipulation/tutorials/group-tags',\n",
       " 'introduction/tutorials/galaxy-intro-peaks2genes',\n",
       " 'statistics/tutorials/machinelearning',\n",
       " 'transcriptomics/tutorials/rna-seq-counts-to-genes',\n",
       " 'transcriptomics/tutorials/rna-seq-viz-with-heatmap2',\n",
       " 'transcriptomics/tutorials/small_ncrna_clustering',\n",
       " 'instructors/tutorials/workshop-intro',\n",
       " 'admin/tutorials/production',\n",
       " 'galaxy-ui/tutorials/workflow-parameters',\n",
       " 'metagenomics/tutorials/mothur-miseq-sop-short',\n",
       " 'computational-chemistry/tutorials/md-simulation-namd',\n",
       " 'computational-chemistry/tutorials/setting-up-molecular-systems',\n",
       " 'galaxy-data-manipulation/tutorials/collections',\n",
       " 'dev/tutorials/visualization-generic',\n",
       " 'galaxy-data-manipulation/tutorials/name-tags',\n",
       " 'dev/tutorials/toolshed',\n",
       " 'transcriptomics/tutorials/rna-seq-genes-to-pathways',\n",
       " 'introduction/tutorials/galaxy-intro-strands',\n",
       " 'contributing/tutorials/create-new-tutorial-technical',\n",
       " 'contributing/tutorials/github-command-line-contribution',\n",
       " 'contributing/tutorials/github-interface-contribution',\n",
       " 'contributing/tutorials/running-jekyll',\n",
       " 'galaxy-ui/tutorials/history-to-workflow',\n",
       " 'genome-annotation/tutorials/genome-annotation',\n",
       " 'galaxy-data-manipulation/tutorials/processing-many-samples-at-once',\n",
       " 'genome-annotation/tutorials/tnseq',\n",
       " 'proteomics/tutorials/labelfree-vs-labelled',\n",
       " 'proteomics/tutorials/ntails',\n",
       " 'admin/tutorials/cloudbursting',\n",
       " 'admin/tutorials/reference-genomes',\n",
       " 'admin/tutorials/troubleshooting',\n",
       " 'contributing/tutorials/create-new-tutorial-slides',\n",
       " 'introduction/tutorials/galaxy-intro-ngs-data-managment',\n",
       " 'variant-analysis/tutorials/non-dip',\n",
       " 'dev/tutorials/containers',\n",
       " 'dev/tutorials/interactive-environments',\n",
       " 'admin/tutorials/galaxy-docker',\n",
       " 'dev/tutorials/interactive-tours',\n",
       " 'dev/tutorials/webhooks',\n",
       " 'galaxy-ui/tutorials/galaxy-intro-jupyter',\n",
       " 'dev/tutorials/architecture',\n",
       " 'admin/tutorials/advanced-galaxy-customisation',\n",
       " 'admin/tutorials/general-monitoring',\n",
       " 'admin/tutorials/maintenance',\n",
       " 'admin/tutorials/tool-install',\n",
       " 'admin/tutorials/upgrading',\n",
       " 'admin/tutorials/stuff',\n",
       " 'admin/tutorials/empathy',\n",
       " 'admin/tutorials/database',\n",
       " 'admin/tutorials/systemd-supervisor',\n",
       " 'admin/tutorials/toolshed',\n",
       " 'admin/tutorials/users-groups-quotas',\n",
       " 'transcriptomics/tutorials/rb-rnaseq',\n",
       " 'ecology/tutorials/species-distribution-modeling',\n",
       " 'contributing/tutorials/generating-pdf',\n",
       " 'contributing/tutorials/create-new-tutorial-tours',\n",
       " 'galaxy-ui/tutorials/history',\n",
       " 'admin/tutorials/database-schema',\n",
       " 'dev/tutorials/conda_sys',\n",
       " 'Dev-Corner/tutorials/files']"
      ]
     },
     "execution_count": 303,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(tuto_prs.keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Statistics per tutorials"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 305,
   "metadata": {},
   "outputs": [],
   "source": [
    "tuto_pr_stats = pd.DataFrame(\n",
    "    0,\n",
    "    columns=['Creation', 'PR', 'Mean duration between PRs'],\n",
    "    index=[])\n",
    "for t in tuto_prs:\n",
    "    tuto_pr_stats.loc[t, 'PR'] = len(tuto_prs[t])\n",
    "    #tuto_pr_stats.loc[t, 'Commits'] = repo.get_commits(path = t).totalCount\n",
    "    # extract mean duration by sorting PRs and compute time between merge and creation\n",
    "    ordered_prs = sorted(tuto_prs[t])\n",
    "    tuto_pr_stats.loc[t, 'Creation'] = prs[ordered_prs[0]].merged_at\n",
    "    prev_pr_merged_at = prs[ordered_prs[0]].merged_at\n",
    "    duration = datetime.timedelta(0)\n",
    "    pr_n = 0\n",
    "    for pr in ordered_prs[1:]:\n",
    "        d = prs[pr].created_at - prev_pr_merged_at\n",
    "        # keep only positive delta\n",
    "        if d < datetime.timedelta(0):\n",
    "            continue\n",
    "        pr_n += 1\n",
    "        duration += d\n",
    "        prev_pr_merged_at = prs[pr].merged_at\n",
    "    if pr_n > 0:\n",
    "        tuto_pr_stats.loc[t, 'Mean duration between PRs'] = duration/pr_n\n",
    "    else:\n",
    "        tuto_pr_stats.loc[t, 'Mean duration between PRs'] = datetime.timedelta(0)\n",
    "tuto_pr_stats.loc[:, 'PR'] = tuto_pr_stats.loc[:, 'PR'].astype(int)\n",
    "tuto_pr_stats = (tuto_pr_stats\n",
    "    .sort_values('PR',ascending=False))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Top 10 tutorials in terms of PR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 308,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Creation</th>\n",
       "      <th>PR</th>\n",
       "      <th>Mean duration between PRs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>transcriptomics/tutorials/ref-based</td>\n",
       "      <td>2016-10-05 15:28:00</td>\n",
       "      <td>77</td>\n",
       "      <td>16 days, 11:22:47.860465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>metagenomics/tutorials/mothur-miseq-sop</td>\n",
       "      <td>2017-02-12 03:33:39</td>\n",
       "      <td>55</td>\n",
       "      <td>17 days, 22:54:26.275000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>epigenetics/tutorials/tal1-binding-site-identification</td>\n",
       "      <td>2016-12-19 09:00:20</td>\n",
       "      <td>49</td>\n",
       "      <td>14 days, 6:37:11.866667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>sequence-analysis/tutorials/mapping</td>\n",
       "      <td>2016-10-04 14:08:30</td>\n",
       "      <td>49</td>\n",
       "      <td>22 days, 12:16:16.027778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>sequence-analysis/tutorials/quality-control</td>\n",
       "      <td>2016-10-04 13:22:53</td>\n",
       "      <td>45</td>\n",
       "      <td>26 days, 20:14:36.275862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>transcriptomics/tutorials/de-novo</td>\n",
       "      <td>2017-02-19 23:07:53</td>\n",
       "      <td>45</td>\n",
       "      <td>19 days, 15:41:11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>introduction/tutorials/galaxy-intro-peaks2genes</td>\n",
       "      <td>2017-05-24 15:51:25</td>\n",
       "      <td>44</td>\n",
       "      <td>15 days, 11:54:29.551724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>epigenetics/tutorials/methylation-seq</td>\n",
       "      <td>2016-10-05 15:07:38</td>\n",
       "      <td>42</td>\n",
       "      <td>21 days, 16:43:24.742857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>contributing/tutorials/create-new-tutorial-content</td>\n",
       "      <td>2017-06-25 22:36:02</td>\n",
       "      <td>42</td>\n",
       "      <td>16 days, 6:53:11.516129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>variant-analysis/tutorials/dip</td>\n",
       "      <td>2016-08-19 14:05:48</td>\n",
       "      <td>40</td>\n",
       "      <td>28 days, 2:46:10.620690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ecology/tutorials/de-novo-rad-seq</td>\n",
       "      <td>2017-02-14 14:09:18</td>\n",
       "      <td>40</td>\n",
       "      <td>28 days, 0:52:02.629630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ecology/tutorials/ref-based-rad-seq</td>\n",
       "      <td>2017-02-14 14:09:18</td>\n",
       "      <td>39</td>\n",
       "      <td>30 days, 23:47:01.272727</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                               Creation  PR  \\\n",
       "transcriptomics/tutorials/ref-based                 2016-10-05 15:28:00  77   \n",
       "metagenomics/tutorials/mothur-miseq-sop             2017-02-12 03:33:39  55   \n",
       "epigenetics/tutorials/tal1-binding-site-identif...  2016-12-19 09:00:20  49   \n",
       "sequence-analysis/tutorials/mapping                 2016-10-04 14:08:30  49   \n",
       "sequence-analysis/tutorials/quality-control         2016-10-04 13:22:53  45   \n",
       "transcriptomics/tutorials/de-novo                   2017-02-19 23:07:53  45   \n",
       "introduction/tutorials/galaxy-intro-peaks2genes     2017-05-24 15:51:25  44   \n",
       "epigenetics/tutorials/methylation-seq               2016-10-05 15:07:38  42   \n",
       "contributing/tutorials/create-new-tutorial-content  2017-06-25 22:36:02  42   \n",
       "variant-analysis/tutorials/dip                      2016-08-19 14:05:48  40   \n",
       "ecology/tutorials/de-novo-rad-seq                   2017-02-14 14:09:18  40   \n",
       "ecology/tutorials/ref-based-rad-seq                 2017-02-14 14:09:18  39   \n",
       "\n",
       "                                                   Mean duration between PRs  \n",
       "transcriptomics/tutorials/ref-based                 16 days, 11:22:47.860465  \n",
       "metagenomics/tutorials/mothur-miseq-sop             17 days, 22:54:26.275000  \n",
       "epigenetics/tutorials/tal1-binding-site-identif...   14 days, 6:37:11.866667  \n",
       "sequence-analysis/tutorials/mapping                 22 days, 12:16:16.027778  \n",
       "sequence-analysis/tutorials/quality-control         26 days, 20:14:36.275862  \n",
       "transcriptomics/tutorials/de-novo                          19 days, 15:41:11  \n",
       "introduction/tutorials/galaxy-intro-peaks2genes     15 days, 11:54:29.551724  \n",
       "epigenetics/tutorials/methylation-seq               21 days, 16:43:24.742857  \n",
       "contributing/tutorials/create-new-tutorial-content   16 days, 6:53:11.516129  \n",
       "variant-analysis/tutorials/dip                       28 days, 2:46:10.620690  \n",
       "ecology/tutorials/de-novo-rad-seq                    28 days, 0:52:02.629630  \n",
       "ecology/tutorials/ref-based-rad-seq                 30 days, 23:47:01.272727  "
      ]
     },
     "execution_count": 308,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tuto_pr_stats.head(12)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Mean number PR and commits per tutorials"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 309,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PR    14.768293\n",
       "dtype: float64"
      ]
     },
     "execution_count": 309,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tuto_pr_stats.mean()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Mean duration between PRs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 315,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'33 days, 17:30:19.207027'"
      ]
     },
     "execution_count": 315,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "str(tuto_pr_stats.loc[:,'Mean duration between PRs'].sum()/tuto_pr_stats.shape[0])"
   ]
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "165px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
